// tdesign-vue-next/esm/table/primary-table/sorter-button.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isVNode, defineComponent, h, createVNode, mergeProps } from 'vue';
import isFunction from 'lodash/isFunction';
import { ChevronDownIcon } from 'tdesign-icons-vue-next';
import mixins from '../../utils/mixins.js';
import getConfigReceiverMixins from '../../config-provider/config-receiver.js';
import { prefix } from '../../config.js';
import { Tooltip } from '../../tooltip/index.js';
import '../../config-provider/zh_CN_config.js';
import '../../config-provider/type.js';
import '../../tooltip/tooltip.js';
import '../../tooltip/props.js';
import '../../popup/props.js';
import '../../popup/index.js';
import '../../popup/popup.js';
import '@babel/runtime/helpers/typeof';
import '@popperjs/core';
import '../../_chunks/dep-b66bfe36.js';
import '../../utils/classnames.js';
import '../../utils/dom.js';
import 'lodash/isString';
import '../../utils/easing.js';
import '../../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isObject';
import '../../utils/set-style.js';
import '../../utils/event.js';
import '../../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../../utils/map-props.js';
import 'lodash/kebabCase';
import '../../utils/withInstall.js';
import '../../popup/style';
import '../../popup/type.js';
import '../../tooltip/style';
import '../../tooltip/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var tooltips = {
  asc: "\u70B9\u51FB\u5347\u5E8F",
  desc: "\u70B9\u51FB\u964D\u5E8F",
  cancel: "\u70B9\u51FB\u53D6\u6D88\u6392\u5E8F"
};
var SorterButton = defineComponent(_objectSpread(_objectSpread({}, mixins(getConfigReceiverMixins("table"))), {}, {
  name: "".concat(prefix, "-sorter-button"),
  props: {
    sortType: {
      type: String,
      "default": null
    },
    sortOrder: {
      type: String,
      "default": function _default() {
        return "";
      }
    },
    nextSortOrder: {
      type: String,
      required: false,
      "default": "cancel"
    }
  },
  emits: ["change"],
  computed: {
    allowSortTypes: function allowSortTypes() {
      return this.sortType === "all" ? ["asc", "desc"] : [this.sortType];
    }
  },
  methods: {
    getSortIcon: function getSortIcon(direction, className) {
      var icon = isFunction(this.global.sortIcon) ? this.global.sortIcon(h) : createVNode(ChevronDownIcon, {
        "size": "16px"
      }, null);
      var style = {
        left: "0px"
      };

      if (direction === "asc") {
        style = _objectSpread({
          transform: "rotate(-180deg)",
          top: "-1px"
        }, style);
      } else {
        style.bottom = "-1px";
      }

      var sortClassName = ["".concat(prefix, "-table__sort-icon"), className, "".concat(prefix, "-table-sort-").concat(direction)];
      return h(icon, {
        style: style,
        "class": sortClassName
      });
    }
  },
  render: function render() {
    var _this = this;

    var allowSortTypes = this.allowSortTypes,
        sortOrder = this.sortOrder,
        nextSortOrder = this.nextSortOrder;
    var buttonProps = {
      "class": allowSortTypes.length > 1 ? "".concat(prefix, "-table__double-icons") : ""
    };
    var tips = tooltips[nextSortOrder];
    var sortButton = allowSortTypes.map(function (direction) {
      var className = direction === sortOrder ? "".concat(prefix, "-table__sort-icon--active") : "".concat(prefix, "-icon-sort--default");
      return _this.getSortIcon(direction, className);
    });
    return createVNode("div", mergeProps({
      "class": "".concat(prefix, "-table__cell--sort-trigger")
    }, buttonProps), [tips ? createVNode(Tooltip, {
      "style": "line-height: 0px;position:relative;",
      "content": tips,
      "showArrow": false
    }, _isSlot(sortButton) ? sortButton : {
      "default": function _default() {
        return [sortButton];
      }
    }) : sortButton]);
  }
}));

export { SorterButton as default };
