// tdesign-vue-next/esm/time-picker/panel/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isVNode, defineComponent, nextTick, createVNode, resolveComponent } from 'vue';
import dayjs from 'dayjs';
import { c as customParseFormat, P as PanelCol } from '../../_chunks/dep-ff8379e8.js';
import mixins from '../../utils/mixins.js';
import getConfigReceiverMixins from '../../config-provider/config-receiver.js';
import { COMPONENT_NAME, EPickerCols } from '../constant.js';
import { panelProps } from './props.js';
import TButton from '../../button/button.js';
import 'lodash/debounce';
import '../../_chunks/dep-b66bfe36.js';
import '../../config.js';
import '../../config-provider/zh_CN_config.js';
import '../../config-provider/type.js';
import '../props.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../../utils/classnames.js';
import '../../loading/index.js';
import '../../loading/loading.js';
import '../../loading/icon/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../utils/dom.js';
import 'lodash/isString';
import '../../utils/easing.js';
import '../../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../../utils/transfer-dom.js';
import '../../loading/props.js';
import '../../utils/withInstall.js';
import '../../loading/style';
import '../../loading/type.js';
import '../../loading/plugin.js';
import '../../button/props.js';
import '../../utils/ripple.js';
import '../../utils/set-style.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var name = "".concat(COMPONENT_NAME, "__panel");
dayjs.extend(customParseFormat);
var PickerPanel = defineComponent(_objectSpread(_objectSpread({}, mixins(getConfigReceiverMixins("timePicker"))), {}, {
  name: name,
  components: {
    PanelCol: PanelCol,
    TButton: TButton
  },
  props: panelProps(),
  emits: ["sure", "now-action", "time-pick"],
  data: function data() {
    return {
      panel: null,
      isSetup: false
    };
  },
  computed: {
    sectionComponentName: function sectionComponentName() {
      return "".concat(name, "-section");
    },
    classNames: function classNames() {
      return this.rangePicker ? [name, this.sectionComponentName] : [name];
    },
    colValues: function colValues() {
      return this.value.map(function (el) {
        return el || dayjs();
      });
    },
    rangePicker: function rangePicker() {
      return this.colValues.length > 1;
    },
    formatField: function formatField() {
      var match = this.format.match(/(a\s+|A\s+)?(h+|H+)?:?(m+)?:?(s+)?(\s+a|\s+A)?/);

      var _match = _slicedToArray(match, 6),
          startAChart = _match[1],
          hour = _match[2],
          minute = _match[3],
          second = _match[4],
          endAChart = _match[5];

      return {
        startAChart: startAChart,
        hour: hour,
        minute: minute,
        second: second,
        endAChart: endAChart
      };
    },
    cols: function cols() {
      if (!this.formatField) {
        return [EPickerCols.hour, EPickerCols.minute, EPickerCols.second];
      }

      var _this$formatField = this.formatField,
          startAChart = _this$formatField.startAChart,
          hour = _this$formatField.hour,
          minute = _this$formatField.minute,
          second = _this$formatField.second,
          endAChart = _this$formatField.endAChart;
      var res = [];
      startAChart && res.push(EPickerCols.meridiem);
      hour && res.push(EPickerCols.hour);
      minute && res.push(EPickerCols.minute);
      second && res.push(EPickerCols.second);
      endAChart && res.push(EPickerCols.meridiem);
      return res;
    },
    localeMeridiems: function localeMeridiems() {
      return [this.global.anteMeridiem, this.global.postMeridiem];
    }
  },
  watch: {
    isShowPanel: {
      handler: function handler(val) {
        if (val) {
          this.panelColUpdate();
        }
      },
      immediate: true
    }
  },
  methods: {
    panelColUpdate: function panelColUpdate() {
      var _this = this;

      nextTick(function () {
        var panelCol0 = _this.$refs.panelCol_0;
        var panelCol1 = _this.$refs.panelCol_1;
        panelCol0 && panelCol0.updateTimeScrollPos();
        panelCol1 && panelCol1.updateTimeScrollPos();
      });
    },
    scrollToTime: function scrollToTime(colIndex, col, time, behavior) {
      var scroller = this.$refs["panelCol_".concat(colIndex)];
      scroller && scroller.scrollToTime(col, time, behavior);
    },
    renderFooter: function renderFooter() {
      var _slot, _slot2;

      var confirmAction = this.confirmBtnClick.bind(this);
      return createVNode("div", {
        "class": "".concat(this.sectionComponentName, "-footer")
      }, [createVNode(resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "base",
        "onClick": confirmAction
      }, _isSlot(_slot = this.t(this.global.confirm)) ? _slot : {
        "default": function _default() {
          return [_slot];
        }
      }), this.rangePicker || createVNode(resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "text",
        "onClick": this.nowAction
      }, _isSlot(_slot2 = this.t(this.global.nowtime)) ? _slot2 : {
        "default": function _default() {
          return [_slot2];
        }
      })]);
    },
    renderBody: function renderBody() {
      return createVNode("div", {
        "class": "".concat(this.sectionComponentName, "-body")
      }, [this.renderSinglePicker(0), this.rangePicker && this.renderSinglePicker(1)]);
    },
    renderSinglePicker: function renderSinglePicker(index) {
      var _this2 = this;

      var val = this.colValues[index];
      var ref = "panelCol_".concat(index);
      return createVNode(resolveComponent("panel-col"), {
        "ref": ref,
        "value": val,
        "cols": this.cols,
        "steps": this.steps,
        "hideDisabledTime": this.hideDisabledTime,
        "disableTime": this.disableTime,
        "format": this.format,
        "onTimePick": function onTimePick(col, time) {
          return _this2.handleTimePick(col, time, index);
        },
        "localeMeridiems": this.localeMeridiems
      }, null);
    },
    confirmBtnClick: function confirmBtnClick() {
      this.$emit("sure");
    },
    nowAction: function nowAction() {
      this.$emit("now-action");
      this.panelColUpdate();
    },
    handleTimePick: function handleTimePick(col, time, index) {
      this.$emit("time-pick", col, time, index, this.colValues[index]);
    }
  },
  render: function render() {
    var isFooterDisplay = this.isFooterDisplay,
        classNames = this.classNames;
    return createVNode("div", {
      "class": classNames
    }, [this.renderBody(), isFooterDisplay ? this.renderFooter() : null]);
  }
}));

export { PickerPanel as default };
