// tdesign-vue-next/lib/table/base-table/col-group.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var vue = require('vue');
var table_baseTableProps = require('../base-table-props.js');
var config = require('../../config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);

var TableColGroup = vue.defineComponent({
  name: "".concat(config.prefix, "-table-col-group"),
  props: {
    columns: table_baseTableProps["default"].columns
  },
  methods: {
    renderColgroup: function renderColgroup() {
      this.columns;
      var colgroup = [];

      var flatArray = function flatArray(arr) {
        var res = [];
        arr.forEach(function (item) {
          var _item$children;

          if ((item === null || item === void 0 ? void 0 : (_item$children = item.children) === null || _item$children === void 0 ? void 0 : _item$children.length) > 0) {
            var val = flatArray(item.children);
            res.push.apply(res, _toConsumableArray__default["default"](val));
          } else {
            res.push(item);
          }
        });
        return res;
      };

      var flatCols = flatArray(this.columns);
      flatCols.forEach(function (column) {
        var width = column.width,
            minWidth = column.minWidth,
            colKey = column.colKey;
        var style = {};

        if (width) {
          style.width = "".concat(width, "px");
        }

        if (minWidth) {
          style.minWidth = "".concat(minWidth, "px");
        }

        colgroup.push(vue.createVNode("col", {
          "key": colKey,
          "style": style
        }, null));
      });
      return colgroup;
    }
  },
  render: function render() {
    return vue.createVNode("colgroup", null, [this.renderColgroup()]);
  }
});

exports["default"] = TableColGroup;
