// tdesign-vue-next/lib/table/base-table/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var throttle = require('lodash/throttle');
var utils_mixins = require('../../utils/mixins.js');
var configProvider_configReceiver = require('../../config-provider/config-receiver.js');
var config = require('../../config.js');
var table_util_propsUtil = require('../util/props-util.js');
var table_baseTableProps = require('../base-table-props.js');
var table_baseTable_tableBody = require('./table-body.js');
var table_baseTable_tableHeader = require('./table-header.js');
var table_baseTable_colGroup = require('./col-group.js');
var pagination_index = require('../../pagination/index.js');
var loading_index = require('../../loading/index.js');
var table_util_common = require('../util/common.js');
var utils_renderTnode = require('../../utils/render-tnode.js');
var table_util_interface = require('../util/interface.js');
var utils_event = require('../../utils/event.js');
var utils_helper = require('../../utils/helper.js');
var utils_classnames = require('../../utils/classnames.js');
var hooks_virtualScroll = require('../../hooks/virtualScroll.js');
require('../../config-provider/zh_CN_config.js');
require('lodash/get');
require('lodash/camelCase');
require('./table-row.js');
require('./table-cell.js');
require('../../popup/index.js');
require('../../popup/popup.js');
require('@babel/runtime/helpers/typeof');
require('@popperjs/core');
require('../../_chunks/dep-eb6b0f94.js');
require('../../utils/dom.js');
require('lodash/isString');
require('../../utils/easing.js');
require('../../popup/props.js');
require('../../utils/set-style.js');
require('../../utils/map-props.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('lodash/kebabCase');
require('../../utils/withInstall.js');
require('../../popup/style');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isObject');
require('@babel/runtime/helpers/slicedToArray');
require('../primary-table-props.js');
require('../../pagination/pagination.js');
require('tdesign-icons-vue-next');
require('../../input-number/index.js');
require('../../input-number/input-number.js');
require('@babel/runtime/helpers/asyncToGenerator');
require('@babel/runtime/regenerator');
require('../../button/index.js');
require('../../button/button.js');
require('../../button/props.js');
require('../../utils/ripple.js');
require('../../loading/loading.js');
require('../../loading/icon/gradient.js');
require('../../_common/js/loading/circle-adapter.js');
require('../../_common/js/utils/set-style.js');
require('../../_common/js/utils/helper.js');
require('../../utils/transfer-dom.js');
require('../../loading/props.js');
require('../../loading/style');
require('../../loading/plugin.js');
require('../../button/style');
require('../../input-number/props.js');
require('../../input-number/style');
require('../../select/index.js');
require('../../select/select.js');
require('lodash/debounce');
require('lodash/set');
require('../../input/index.js');
require('../../input/addon.js');
require('../../input/input.js');
require('../../input/props.js');
require('../../input/input-group.js');
require('../../input/style');
require('../../tag/index.js');
require('../../tag/tag.js');
require('../../tag/props.js');
require('../../tag/check-tag.js');
require('../../tag/check-tag-props.js');
require('../../tag/style');
require('../../common-components/fake-arrow.js');
require('../../select/option.js');
require('../../select/option-props.js');
require('../../checkbox/index.js');
require('../../checkbox/checkbox.js');
require('../../checkbox/props.js');
require('../../checkbox/group.js');
require('lodash/intersection');
require('../../checkbox/checkbox-group-props.js');
require('../../checkbox/style');
require('../../select/props.js');
require('../../select/optionGroup.js');
require('../../select/option-group-props.js');
require('../../select/style');
require('../../pagination/props.js');
require('../../pagination/style');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var throttle__default = /*#__PURE__*/_interopDefaultLegacy(throttle);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var _BaseTable = vue.defineComponent(_objectSpread(_objectSpread({}, utils_mixins["default"](configProvider_configReceiver["default"]("table"))), {}, {
  name: "TBaseTable",
  components: {
    TableBody: table_baseTable_tableBody["default"],
    TableHeader: table_baseTable_tableHeader["default"],
    TableColGroup: table_baseTable_colGroup["default"],
    Pagination: pagination_index.Pagination
  },
  props: _objectSpread(_objectSpread({}, table_baseTableProps["default"]), {}, {
    onRowDragover: Function,
    onRowDragstart: Function,
    provider: {
      type: Object,
      "default": function _default() {
        return {};
      }
    }
  }),
  emits: ["page-change", "scroll-x", "scroll-y"].concat(_toConsumableArray__default["default"](table_util_interface.EVENT_NAME_WITH_KEBAB)),
  setup: function setup(props) {
    var scrollBody = vue.ref(null);
    vue.provide("scrollBody", scrollBody);

    var _ref = props.scroll || {},
        type = _ref.type,
        rowHeight = _ref.rowHeight,
        _ref$bufferSize = _ref.bufferSize,
        bufferSize = _ref$bufferSize === void 0 ? 20 : _ref$bufferSize,
        _ref$isFixedRowHeight = _ref.isFixedRowHeight,
        isFixedRowHeight = _ref$isFixedRowHeight === void 0 ? false : _ref$isFixedRowHeight;

    var _toRefs = vue.toRefs(props),
        data = _toRefs.data;

    var _ref2 = type === "virtual" ? hooks_virtualScroll["default"]({
      container: scrollBody,
      data: data,
      fixedHeight: isFixedRowHeight,
      lineHeight: rowHeight,
      bufferSize: bufferSize
    }) : {},
        _ref2$trs = _ref2.trs,
        trs = _ref2$trs === void 0 ? null : _ref2$trs,
        _ref2$scrollHeight = _ref2.scrollHeight,
        scrollHeight = _ref2$scrollHeight === void 0 ? null : _ref2$scrollHeight,
        _ref2$visibleData = _ref2.visibleData,
        visibleData = _ref2$visibleData === void 0 ? null : _ref2$visibleData,
        _ref2$translateY = _ref2.translateY,
        translateY = _ref2$translateY === void 0 ? null : _ref2$translateY,
        _ref2$handleScroll = _ref2.handleScroll,
        handleScroll = _ref2$handleScroll === void 0 ? null : _ref2$handleScroll,
        _ref2$handleRowMounte = _ref2.handleRowMounted,
        handleRowMounted = _ref2$handleRowMounte === void 0 ? null : _ref2$handleRowMounte;

    return {
      scrollType: type,
      rowHeight: rowHeight,
      trs: trs,
      bufferSize: bufferSize,
      scrollBody: scrollBody,
      scrollHeight: scrollHeight,
      visibleData: visibleData,
      translateY: translateY,
      handleRowMounted: handleRowMounted,
      handleVirtualScroll: handleScroll
    };
  },
  data: function data() {
    return {
      scrollableToLeft: false,
      scrollableToRight: false,
      scrollBarWidth: 0,
      defaultCurrent: 0,
      defaultPageSize: 0,
      useFixedHeader: false
    };
  },
  computed: {
    current: function current() {
      var _this$pagination, _this$pagination2;

      return ((_this$pagination = this.pagination) === null || _this$pagination === void 0 ? void 0 : _this$pagination.current) || this.defaultCurrent || ((_this$pagination2 = this.pagination) === null || _this$pagination2 === void 0 ? void 0 : _this$pagination2.defaultCurrent);
    },
    pageSize: function pageSize() {
      var _this$pagination3, _this$pagination4;

      return ((_this$pagination3 = this.pagination) === null || _this$pagination3 === void 0 ? void 0 : _this$pagination3.pageSize) || this.defaultPageSize || ((_this$pagination4 = this.pagination) === null || _this$pagination4 === void 0 ? void 0 : _this$pagination4.defaultPageSize);
    },
    dataSource: function dataSource() {
      if (!this.hasPagination) return this.data.slice(0);
      var current = this.current,
          pageSize = this.pageSize;

      if (this.data.length > pageSize) {
        return this.data.slice((current - 1) * pageSize, current * pageSize);
      }

      return this.data;
    },
    flattedColumns: function flattedColumns() {
      return table_util_propsUtil.flatColumns(this.columns);
    },
    isEmpty: function isEmpty() {
      return (!this.dataSource || this.dataSource.length === 0) && !this.loading;
    },
    hasFixedColumns: function hasFixedColumns() {
      var columns = this.columns;
      return columns.some(function (item) {
        return item.fixed === "right" || item.fixed === "left";
      });
    },
    hasPagination: function hasPagination() {
      return !!this.pagination;
    },
    isLoading: function isLoading() {
      return !!this.loading;
    },
    tableHeight: function tableHeight() {
      var height = this.height,
          maxHeight = this.maxHeight,
          useFixedHeader = this.useFixedHeader,
          isEmpty = this.isEmpty;

      if (isEmpty) {
        return "auto";
      }

      if (height !== "auto" && height) {
        return height;
      }

      if (maxHeight && useFixedHeader) {
        return maxHeight;
      }

      return "auto";
    },
    fixedHeader: function fixedHeader() {
      var tableHeight = this.tableHeight;
      return tableHeight !== "auto";
    },
    commonClass: function commonClass() {
      var _ref3;

      var classes = ["".concat(config.prefix, "-table"), (_ref3 = {}, _defineProperty__default["default"](_ref3, utils_classnames.SIZE_CLASSNAMES.small, this.size === "small"), _defineProperty__default["default"](_ref3, utils_classnames.SIZE_CLASSNAMES.large, this.size === "large"), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table--bordered"), this.bordered), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table--striped"), this.stripe), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table--hoverable"), this.hover), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table__row--draggable"), this.provider.sortOnRowDraggable), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table-table--align-top"), this.verticalAlign === "top"), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table-table--align-bottom"), this.verticalAlign === "bottom"), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table__cell--fixed"), this.hasFixedColumns), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table--has-fixed"), this.hasFixedColumns), _defineProperty__default["default"](_ref3, "".concat(config.prefix, "-table__header--fixed"), this.fixedHeader), _ref3)];
      return classes;
    },
    usePadding: function usePadding() {
      return this.fixedHeader || this.scrollableToRight || this.scrollableToLeft;
    }
  },
  mounted: function mounted() {
    var _this = this;

    if (this.hasFixedColumns) {
      setTimeout(function () {
        _this.checkScrollableToLeftOrRight();
      }, 0);
      this.addWindowResizeEventListener();
    }

    var scrollDiv = document.createElement("div");
    scrollDiv.style.cssText = "\n      width: 99px;\n      height: 99px;\n      overflow: scroll;\n      position: absolute;\n      top: -9999px;";
    scrollDiv.classList.add("scrollbar");
    document.body.appendChild(scrollDiv);
    this.scrollBarWidth = scrollDiv.offsetWidth - scrollDiv.clientWidth;
    document.body.removeChild(scrollDiv);
    this.maxHeight;
    this.checkMaxHeight();
  },
  unmounted: function unmounted() {
    window.removeEventListener("resize", table_util_common.debounce(this.checkScrollableToLeftOrRight));
  },
  updated: function updated() {
    this.checkMaxHeight();
  },
  methods: {
    checkScrollableToLeftOrRight: function checkScrollableToLeftOrRight() {
      var scrollContainer = this.$refs[this.fixedHeader ? "scrollBody" : "tableContent"];
      if (!scrollContainer) return;
      var scrollLeft = scrollContainer.scrollLeft,
          scrollWidth = scrollContainer.scrollWidth,
          clientWidth = scrollContainer.clientWidth;
      this.scrollableToLeft = scrollLeft > 0;
      this.scrollableToRight = scrollLeft + clientWidth < scrollWidth;
    },
    addWindowResizeEventListener: function addWindowResizeEventListener() {
      window.addEventListener("resize", table_util_common.debounce(this.checkScrollableToLeftOrRight));
    },
    renderHeader: function renderHeader() {
      var columns = this.columns,
          $slots = this.$slots,
          bordered = this.bordered;
      return vue.createVNode(table_baseTable_tableHeader["default"], {
        "columns": columns,
        "bordered": bordered
      }, $slots);
    },
    renderBody: function renderBody() {
      var _this2 = this;

      var $slots = this.$slots;
      var rowEvents = {};
      table_util_interface.EVENT_NAME_WITH_KEBAB.concat(["row-dragstart", "row-dragover"]).forEach(function (eventName) {
        rowEvents[utils_helper.getPropsApiByEvent(eventName)] = function (params) {
          utils_event.emitEvent(_this2, eventName, params);
        };
      });
      var props = {
        rowKey: this.rowKey,
        data: this.scrollType === "virtual" ? this.visibleData : this.dataSource,
        provider: this.provider,
        columns: this.flattedColumns,
        rowClassName: this.rowClassName,
        current: this.current,
        rowspanAndColspan: this.rowspanAndColspan,
        firstFullRow: this.firstFullRow,
        lastFullRow: this.lastFullRow,
        scrollType: this.scrollType,
        rowHeight: this.rowHeight,
        trs: this.trs,
        bufferSize: this.bufferSize,
        handleRowMounted: this.handleRowMounted
      };
      return vue.createVNode(table_baseTable_tableBody["default"], vue.mergeProps(props, rowEvents), _isSlot($slots) ? $slots : {
        "default": function _default() {
          return [$slots];
        }
      });
    },
    renderEmptyTable: function renderEmptyTable() {
      var useLocale = !this.empty && !this.$slots.empty;
      var height = this.height;
      var wrapperStyle = {};

      if (height !== "auto") {
        wrapperStyle.height = isNaN(Number(height)) ? height : "".concat(height, "px");
      }

      return vue.createVNode("div", {
        "style": wrapperStyle,
        "class": "".concat(config.prefix, "-table__empty")
      }, [useLocale ? this.global.empty : utils_renderTnode.renderTNodeJSX(this, "empty")]);
    },
    renderPagination: function renderPagination() {
      var _this3 = this;

      var paginationProps = this.pagination;
      return vue.createVNode("div", {
        "class": "".concat(config.prefix, "-table__pagination")
      }, [vue.createVNode(pagination_index.Pagination, vue.mergeProps(paginationProps, {
        onChange: function onChange(pageInfo) {
          paginationProps.onChange && paginationProps.onChange(pageInfo);
        },
        onCurrentChange: function onCurrentChange(current, pageInfo) {
          utils_event.emitEvent(_this3, "page-change", pageInfo, _this3.dataSource);
          _this3.defaultCurrent = current;
          paginationProps.onCurrentChange && paginationProps.onCurrentChange(current, pageInfo);
        },
        onPageSizeChange: function onPageSizeChange(pageSize, pageInfo) {
          utils_event.emitEvent(_this3, "page-change", pageInfo, _this3.dataSource);
          _this3.defaultPageSize = pageSize;
          paginationProps.onPageSizeChange && paginationProps.onPageSizeChange(pageSize, pageInfo);
        }
      }), null)]);
    },
    renderTableWithFixedHeader: function renderTableWithFixedHeader() {
      var _this4 = this;

      var fixedTable = [];
      var columns = this.columns,
          asyncLoadingProps = this.provider.asyncLoadingProps,
          tableLayout = this.tableLayout,
          scrollBarWidth = this.scrollBarWidth,
          hasFixedColumns = this.hasFixedColumns,
          tableHeight = this.tableHeight,
          usePadding = this.usePadding;
      var handleScroll = throttle__default["default"](function (e) {
        var target = e.target;
        var scrollLeft = target.scrollLeft;
        _this4.$refs.scrollHeader.scrollLeft = scrollLeft;

        _this4.handleScroll(e);

        _this4.scrollType === "virtual" && _this4.handleVirtualScroll();
      }, 10);
      var paddingRight = "".concat(scrollBarWidth, "px");
      var headerContainerStyle = columns.length > 1 && usePadding ? {
        paddingRight: paddingRight
      } : {};
      fixedTable.push(vue.createVNode("div", {
        "class": "".concat(config.prefix, "-table__header"),
        "style": headerContainerStyle,
        "ref": "scrollHeader"
      }, [vue.createVNode("table", {
        "style": {
          tableLayout: tableLayout
        }
      }, [vue.createVNode(table_baseTable_colGroup["default"], {
        "columns": columns
      }, null), this.renderHeader()])]));
      var containerStyle = {
        height: isNaN(Number(tableHeight)) ? tableHeight : "".concat(Number(tableHeight), "px"),
        width: hasFixedColumns ? "100%" : void 0,
        position: "relative",
        overscrollBehavior: "none"
      };
      var isVirtual = this.scrollType === "virtual";
      fixedTable.push(vue.createVNode("div", vue.mergeProps({
        "class": "".concat(config.prefix, "-table__body"),
        "style": containerStyle
      }, asyncLoadingProps, {
        "ref": "scrollBody",
        "onScroll": handleScroll
      }), [isVirtual && vue.createVNode("div", {
        "style": {
          position: "absolute",
          width: "1px",
          height: "1px",
          transition: "transform .2s",
          transform: "translate(0, ".concat(this.scrollHeight, "px)")
        }
      }, null), vue.createVNode("table", {
        "ref": "table",
        "style": {
          tableLayout: tableLayout,
          transform: isVirtual && "translate(0, ".concat(this.translateY, "px)")
        }
      }, [vue.createVNode(table_baseTable_colGroup["default"], {
        "columns": columns
      }, null), this.renderBody(), this.renderFooter()])]));
      return fixedTable;
    },
    renderLoadingContent: function renderLoadingContent() {
      return utils_renderTnode.renderTNodeJSX(this, "loading", vue.createVNode("div", null, null));
    },
    renderFooter: function renderFooter() {
      var colspan = this.flattedColumns.length;
      var footerContent = utils_renderTnode.renderTNodeJSX(this, "footer");
      return footerContent ? vue.createVNode("tfoot", null, [vue.createVNode("tr", null, [vue.createVNode("td", {
        "colspan": colspan
      }, [footerContent])])]) : null;
    },
    handleScroll: function handleScroll(e) {
      this.checkScrollableToLeftOrRight();
      var _e$target = e.target,
          scrollLeft = _e$target.scrollLeft,
          scrollTop = _e$target.scrollTop;
      var direction = table_util_common.getScrollDirection(scrollLeft, scrollTop);

      if (direction !== table_util_common.ScrollDirection.UNKNOWN) {
        var scrollListenerName = direction === table_util_common.ScrollDirection.X ? "scroll-x" : "scroll-y";
        utils_event.emitEvent(this, scrollListenerName, {
          e: e
        });
      }
    },
    checkMaxHeight: function checkMaxHeight() {
      var maxHeight = this.maxHeight;

      if (maxHeight && this.$refs.tableContent.clientHeight > maxHeight) {
        this.useFixedHeader = true;
      }
    }
  },
  render: function render() {
    var _ref4;

    var hasPagination = this.hasPagination,
        commonClass = this.commonClass,
        fixedHeader = this.fixedHeader,
        columns = this.columns,
        tableLayout = this.tableLayout,
        isLoading = this.isLoading,
        isEmpty = this.isEmpty,
        useFixedHeader = this.useFixedHeader,
        hasFixedColumns = this.hasFixedColumns;
    var body = [];

    var tableColGroup = vue.createVNode(table_baseTable_colGroup["default"], {
      "columns": columns
    }, null);

    var tableHeader = this.renderHeader();
    var tableContent = [tableColGroup, tableHeader];
    var fixedTableContent;

    if (fixedHeader || useFixedHeader) {
      fixedTableContent = this.renderTableWithFixedHeader();
    } else {
      tableContent.push(this.renderBody());
      tableContent.push(this.renderFooter());
    }

    if (isEmpty) {
      body.push(this.renderEmptyTable());
    }

    if (hasPagination) {
      body.push(this.renderPagination());
    }

    var handleScroll = throttle__default["default"](this.handleScroll, 100);
    var tableContentClass = ["".concat(config.prefix, "-table__content"), (_ref4 = {}, _defineProperty__default["default"](_ref4, "".concat(config.prefix, "-table__content--scrollable-to-right"), this.scrollableToRight), _defineProperty__default["default"](_ref4, "".concat(config.prefix, "-table__content--scrollable-to-left"), this.scrollableToLeft), _ref4)];
    var width;
    var _this$$refs = this.$refs,
        tableContentEl = _this$$refs.tableContent,
        tableEl = _this$$refs.table;

    if (!hasFixedColumns && (tableContentEl === null || tableContentEl === void 0 ? void 0 : tableContentEl.clientWidth) < (tableEl === null || tableEl === void 0 ? void 0 : tableEl.clientWidth)) {
      width = "".concat(tableEl.clientWidth, "px");
    }

    return vue.createVNode("div", {
      "class": commonClass,
      "style": {
        width: width
      }
    }, [vue.createVNode(loading_index.Loading, {
      "loading": isLoading,
      "showOverlay": true,
      "text": this.renderLoadingContent
    }, {
      "default": function _default() {
        return [vue.createVNode("div", {
          "ref": "tableContent",
          "class": tableContentClass,
          "onScroll": handleScroll
        }, [fixedTableContent || vue.createVNode("table", {
          "ref": "table",
          "style": {
            tableLayout: tableLayout
          }
        }, [tableContent])]), body];
      }
    })]);
  }
}));

exports["default"] = _BaseTable;
