// tdesign-vue-next/lib/tree-select/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  clearable: Boolean,
  collapsedItems: {
    type: Function
  },
  data: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  disabled: Boolean,
  empty: {
    type: [String, Function],
    "default": ""
  },
  filter: {
    type: Function
  },
  filterable: Boolean,
  loading: Boolean,
  loadingText: {
    type: [String, Function],
    "default": ""
  },
  max: {
    type: Number,
    "default": 0
  },
  minCollapsedNum: {
    type: Number,
    "default": 0
  },
  multiple: Boolean,
  placeholder: {
    type: String,
    "default": ""
  },
  popupProps: {
    type: Object
  },
  prefixIcon: {
    type: Function
  },
  showCheckAlll: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  treeProps: {
    type: Object
  },
  value: {
    type: [String, Number, Object, Array]
  },
  defaultValue: {
    type: [String, Number, Object, Array]
  },
  valueDisplay: {
    type: Function
  },
  valueType: {
    type: String,
    "default": "value",
    validator: function validator(val) {
      return ["value", "object"].includes(val);
    }
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onFocus: Function,
  onRemove: Function,
  onSearch: Function
};

exports["default"] = props;
