// tdesign-vue-next/es/_chunks/dep-3c1f9b23.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import { Button } from '../button/index.js';
import { Divider } from '../divider/index.js';
import { Icon } from '../icon/index.js';
import { Row, Col } from '../grid/index.js';
import { Aside, Layout, Header, Footer, Content } from '../layout/index.js';
import { Affix } from '../affix/index.js';
import { Anchor, AnchorItem, AnchorTarget } from '../anchor/index.js';
import { Breadcrumb, BreadcrumbItem } from '../breadcrumb/index.js';
import { Dropdown, DropdownItem, DropdownMenu } from '../dropdown/index.js';
import { Menu, HeadMenu, Submenu, MenuItem, MenuGroup } from '../menu/index.js';
import { Pagination } from '../pagination/index.js';
import { Steps, StepItem } from '../steps/index.js';
import { Tabs, TabPanel } from '../tabs/index.js';
import { Cascader } from '../cascader/index.js';
import { Checkbox, CheckboxGroup } from '../checkbox/index.js';
import { DatePicker } from '../date-picker/index.js';
import { Form, FormItem } from '../form/index.js';
import { Addon, Input, InputGroup } from '../input/index.js';
import { InputNumber } from '../input-number/index.js';
import { Radio, RadioGroup, RadioButton } from '../radio/index.js';
import { Select, Option, OptionGroup } from '../select/index.js';
import { Slider, SliderButton } from '../slider/index.js';
import { Switch } from '../switch/index.js';
import { TagInput } from '../tag-input/index.js';
import { Textarea } from '../textarea/index.js';
import { Transfer } from '../transfer/index.js';
import { TimePicker, TimeRangePicker } from '../time-picker/index.js';
import { TreeSelect } from '../tree-select/index.js';
import { Avatar, AvatarGroup } from '../avatar/index.js';
import { Badge } from '../badge/index.js';
import { Calendar } from '../calendar/index.js';
import { Comment } from '../comment/index.js';
import { List, ListItem, ListItemMeta } from '../list/index.js';
import { Progress } from '../progress/index.js';
import { Skeleton } from '../skeleton/index.js';
import { Swiper, SwiperItem } from '../swiper/index.js';
import { BaseTable, PrimaryTable, EnhancedTable } from '../table/index.js';
import { Tag, CheckTag } from '../tag/index.js';
import { Tooltip } from '../tooltip/index.js';
import { Tree } from '../tree/index.js';
import { Alert } from '../alert/index.js';
import { Dialog } from '../dialog/index.js';
import { Drawer } from '../drawer/index.js';
import { Loading } from '../loading/index.js';
import { Message } from '../message/index.js';
import { Notification } from '../notification/index.js';
import { Popconfirm } from '../popconfirm/index.js';
import { Popup } from '../popup/index.js';
import { Upload } from '../upload/index.js';
import { ConfigProvider } from '../config-provider/index.js';
import DialogPlugin from '../dialog/plugin.js';
import LoadingPlugin from '../loading/plugin.js';
import MessagePlugin from '../message/plugin.js';
import NotificationPlugin from '../notification/plugin.js';
import { useReceiver } from '../config-provider/useReceiver.js';

var components = /*#__PURE__*/Object.freeze({
	__proto__: null,
	Button: Button,
	Divider: Divider,
	Icon: Icon,
	Row: Row,
	Col: Col,
	Aside: Aside,
	Layout: Layout,
	Header: Header,
	Footer: Footer,
	Content: Content,
	Affix: Affix,
	Anchor: Anchor,
	AnchorItem: AnchorItem,
	AnchorTarget: AnchorTarget,
	Breadcrumb: Breadcrumb,
	BreadcrumbItem: BreadcrumbItem,
	Dropdown: Dropdown,
	DropdownItem: DropdownItem,
	DropdownMenu: DropdownMenu,
	Menu: Menu,
	HeadMenu: HeadMenu,
	Submenu: Submenu,
	MenuItem: MenuItem,
	MenuGroup: MenuGroup,
	Pagination: Pagination,
	Steps: Steps,
	StepItem: StepItem,
	Tabs: Tabs,
	TabPanel: TabPanel,
	Cascader: Cascader,
	Checkbox: Checkbox,
	CheckboxGroup: CheckboxGroup,
	DatePicker: DatePicker,
	Form: Form,
	FormItem: FormItem,
	Addon: Addon,
	Input: Input,
	InputGroup: InputGroup,
	InputNumber: InputNumber,
	Radio: Radio,
	RadioGroup: RadioGroup,
	RadioButton: RadioButton,
	Select: Select,
	Option: Option,
	OptionGroup: OptionGroup,
	Slider: Slider,
	SliderButton: SliderButton,
	Switch: Switch,
	TagInput: TagInput,
	Textarea: Textarea,
	Transfer: Transfer,
	TimePicker: TimePicker,
	TimeRangePicker: TimeRangePicker,
	TreeSelect: TreeSelect,
	Avatar: Avatar,
	AvatarGroup: AvatarGroup,
	Badge: Badge,
	Calendar: Calendar,
	Comment: Comment,
	List: List,
	ListItem: ListItem,
	ListItemMeta: ListItemMeta,
	Progress: Progress,
	Skeleton: Skeleton,
	Swiper: Swiper,
	SwiperItem: SwiperItem,
	BaseTable: BaseTable,
	PrimaryTable: PrimaryTable,
	EnhancedTable: EnhancedTable,
	Tag: Tag,
	CheckTag: CheckTag,
	Tooltip: Tooltip,
	Tree: Tree,
	Alert: Alert,
	DialogPlugin: DialogPlugin,
	Dialog: Dialog,
	Drawer: Drawer,
	LoadingPlugin: LoadingPlugin,
	Loading: Loading,
	MessagePlugin: MessagePlugin,
	Message: Message,
	NotifyPlugin: NotificationPlugin,
	Notification: Notification,
	Popconfirm: Popconfirm,
	Popup: Popup,
	Upload: Upload,
	useReceiver: useReceiver,
	ConfigProvider: ConfigProvider
});

export { components as c };
