// tdesign-vue-next/esm/cascader/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

var props = {
  checkProps: {
    type: Object
  },
  checkStrictly: Boolean,
  clearable: Boolean,
  collapsedItems: {
    type: Function
  },
  disabled: Boolean,
  empty: {
    type: [String, Function]
  },
  filterable: Boolean,
  keys: {
    type: Object
  },
  lazy: {
    type: Boolean,
    "default": true
  },
  load: {
    type: Function
  },
  loading: Boolean,
  loadingText: {
    type: [String, Function],
    "default": ""
  },
  max: {
    type: Number,
    "default": 0
  },
  minCollapsedNum: {
    type: Number,
    "default": 0
  },
  multiple: Boolean,
  options: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  popupProps: {
    type: Object
  },
  showAllLevels: {
    type: Boolean,
    "default": true
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["large", "medium", "small"].includes(val);
    }
  },
  trigger: {
    type: String,
    "default": "click",
    validator: function validator(val) {
      return ["click", "hover"].includes(val);
    }
  },
  value: {
    type: [String, Number, Array],
    "default": function _default() {
      return [];
    }
  },
  defaultValue: {
    type: [String, Number, Array],
    "default": function _default() {
      return [];
    }
  },
  valueMode: {
    type: String,
    "default": "onlyLeaf",
    validator: function validator(val) {
      return ["onlyLeaf", "parentFirst", "all"].includes(val);
    }
  },
  valueType: {
    type: String,
    "default": "single",
    validator: function validator(val) {
      return ["single", "full"].includes(val);
    }
  },
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  onRemove: Function
};

export { props as default };
