// tdesign-vue-next/lib/message/const.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var THEME_LIST = ["info", "success", "warning", "error", "question", "loading"];
var DISTANCE = "32px";
var PLACEMENT_OFFSET = {
  top: {
    top: DISTANCE,
    left: "50%",
    transform: "translateX(-50%)"
  },
  center: {
    left: "50%",
    top: "50%",
    transform: "translateX(-50%) translateY(-50%)"
  },
  left: {
    left: DISTANCE,
    top: "50%",
    transform: "translateY(-50%)"
  },
  bottom: {
    bottom: DISTANCE,
    left: "50%",
    transform: "translateX(-50%)"
  },
  right: {
    right: DISTANCE,
    top: "50%",
    transform: "translateY(-50%)"
  },
  "top-left": {
    left: DISTANCE,
    top: DISTANCE
  },
  "top-right": {
    right: DISTANCE,
    top: DISTANCE
  },
  "bottom-right": {
    right: DISTANCE,
    bottom: DISTANCE
  },
  "bottom-left": {
    left: DISTANCE,
    bottom: DISTANCE
  }
};
var PLACEMENT_LIST = Object.keys(PLACEMENT_OFFSET);
var _const = {
  THEME_LIST: THEME_LIST
};

exports.PLACEMENT_LIST = PLACEMENT_LIST;
exports.PLACEMENT_OFFSET = PLACEMENT_OFFSET;
exports.THEME_LIST = THEME_LIST;
exports["default"] = _const;
