// tdesign-vue-next/lib/input/addon.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var name = "".concat(config.prefix, "-addon");
var _Addon = vue.defineComponent({
  name: name,
  inheritAttrs: false,
  props: {
    prepend: {
      type: [String, Function],
      "default": function _default() {
        return "";
      }
    },
    append: {
      type: [String, Function],
      "default": function _default() {
        return "";
      }
    }
  },
  methods: {
    renderAddon: function renderAddon(h2, type, addon) {
      var addonNode;

      if (this.$slots[type]) {
        addonNode = this.$slots[type](null);
      } else if (typeof addon === "string") {
        addonNode = addon;
      } else if (typeof addon === "function") {
        addonNode = addon(h2);
      } else {
        addonNode = null;
      }

      return addonNode ? vue.createVNode("span", {
        "class": "".concat(name, "__").concat(type)
      }, [addonNode]) : addonNode;
    }
  },
  render: function render() {
    var _ref;

    var prepend = this.renderAddon(vue.h, "prepend", this.prepend);
    var append = this.renderAddon(vue.h, "append", this.append);
    var defaultSlot = this.$slots["default"] ? this.$slots["default"](null) : [null];
    var className = [name, (_ref = {}, _defineProperty__default["default"](_ref, "".concat(name, "--prepend"), prepend), _defineProperty__default["default"](_ref, "".concat(name, "--append"), append), _ref)];

    if (!prepend && !append) {
      return defaultSlot[0];
    }

    return vue.createVNode("div", {
      "class": className
    }, [prepend, defaultSlot[0], append]);
  }
});

exports["default"] = _Addon;
