// tdesign-vue-next/lib/input/input-group.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var name = "".concat(config.prefix, "-input-group");
var _InputGroup = vue.defineComponent({
  name: "TInputGroup",
  props: {
    separate: {
      type: Boolean,
      "default": false
    }
  },
  computed: {
    CLASS: function CLASS() {
      return [name, _defineProperty__default["default"]({}, "".concat(name, "--separate"), this.separate)];
    }
  },
  render: function render() {
    var CLASS = this.CLASS;
    return vue.createVNode("div", {
      "class": CLASS
    }, [this.$slots["default"] && this.$slots["default"](null)]);
  }
});

exports["default"] = _InputGroup;
