// tdesign-vue-next/es/radio/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _Radio from './radio.js';
import _Group from './group.js';
import _RadioButton from './radio-button.js';
import mapProps from '../utils/map-props.js';
import { withInstall } from '../utils/withInstall.js';
import './style/css.js';
import './type.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/defineProperty';
import 'vue';
import '../config.js';
import '../utils/classnames.js';
import '../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import './props.js';
import '../utils/event.js';
import 'lodash/isString';
import 'lodash/isNumber';
import './radio-group-props.js';
import 'lodash/kebabCase';

var Radio = withInstall(mapProps([{
  name: "checked",
  event: "change",
  alias: ["modelValue"]
}])(_Radio));
var RadioGroup = withInstall(mapProps([{
  name: "value",
  event: "change",
  alias: ["modelValue"]
}])(_Group));
var RadioButton = withInstall(_RadioButton);

export { Radio, RadioButton, RadioGroup, Radio as default };
