// tdesign-vue-next/esm/avatar/group.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isVNode, defineComponent, provide, createVNode } from 'vue';
import { prefix } from '../config.js';
import props from './avatar-group-props.js';
import _Avatar from './avatar.js';
import { renderContent, renderTNodeJSX } from '../utils/render-tnode.js';
import '../hooks/event.js';
import '../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../utils/classnames.js';
import './props.js';
import 'lodash/isEmpty';
import 'lodash/isString';
import 'lodash/isFunction';
import 'lodash/isObject';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var name = "".concat(prefix, "-avatar-group");
var _AvatarGroup = defineComponent({
  name: "TAvatarGroup",
  components: {
    Avatar: _Avatar
  },
  props: props,
  setup: function setup(props2) {
    provide("avatarGroup", _objectSpread({}, props2));

    var renderIcon = function renderIcon(context) {
      return isIcon(context) && typeof props2.collapseAvatar !== "string" ? props2.collapseAvatar : null;
    };

    var renderEllipsisAvatar = function renderEllipsisAvatar(context, children) {
      if ((children === null || children === void 0 ? void 0 : children.length) > props2.max) {
        var content = setEllipsisContent(context, children);
        var outAvatar = children.slice(0, props2.max);
        outAvatar.push(createVNode(_Avatar, {
          "size": props2.size,
          "icon": renderIcon(context)
        }, _isSlot(content) ? content : {
          "default": function _default() {
            return [content];
          }
        }));
        return [outAvatar];
      }

      return [children];
    };

    var setEllipsisContent = function setEllipsisContent(context, children) {
      var content = "";

      if (props2.collapseAvatar) {
        if (!isIcon(context)) {
          content = renderContent(context, "collapseAvatar", "content");
        }
      } else {
        content = "+".concat(children.length - props2.max);
      }

      return content;
    };

    var isIcon = function isIcon(context) {
      var content = renderTNodeJSX(context, "collapseAvatar");
      return content;
    };

    return {
      renderEllipsisAvatar: renderEllipsisAvatar,
      isIcon: isIcon,
      setEllipsisContent: setEllipsisContent
    };
  },
  render: function render() {
    var _ref;

    var children = renderTNodeJSX(this, "default");
    var _this$$props = this.$props,
        cascading = _this$$props.cascading,
        max = _this$$props.max;
    var groupClass = ["".concat(name), (_ref = {}, _defineProperty(_ref, "".concat(prefix, "-avatar--offset-right"), cascading === "right-up"), _defineProperty(_ref, "".concat(prefix, "-avatar--offset-left"), cascading === "left-up"), _ref)];
    var content = [children];

    if (max && max >= 0) {
      content = [this.renderEllipsisAvatar(this, children)];
    }

    return createVNode("div", {
      "class": groupClass
    }, [content]);
  }
});

export { _AvatarGroup as default };
