// tdesign-vue-next/esm/tag-input/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

var props = {
  clearable: Boolean,
  collapsedItems: {
    type: Function
  },
  disabled: Boolean,
  dragSort: Boolean,
  excessTagsDisplayType: {
    type: String,
    "default": "scroll",
    validator: function validator(val) {
      return ["scroll", "break-line"].includes(val);
    }
  },
  inputProps: {
    type: Object
  },
  label: {
    type: [String, Function]
  },
  max: {
    type: Number
  },
  minCollapsedNum: {
    type: Number,
    "default": 0
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  readonly: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  status: {
    type: String,
    validator: function validator(val) {
      return ["success", "warning", "error"].includes(val);
    }
  },
  suffix: {
    type: [String, Function]
  },
  suffixIcon: {
    type: Function
  },
  tag: {
    type: [String, Function]
  },
  tagProps: {
    type: Object
  },
  tips: {
    type: [String, Function]
  },
  value: {
    type: Array
  },
  modelValue: {
    type: Array
  },
  defaultValue: {
    type: Array
  },
  valueDisplay: {
    type: [String, Function]
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onEnter: Function,
  onFocus: Function,
  onMouseenter: Function,
  onMouseleave: Function,
  onPaste: Function,
  onRemove: Function
};

export { props as default };
