// tdesign-vue-next/es/input/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _Addon from './addon.js';
import _Input from './input.js';
import _InputGroup from './input-group.js';
import mapProps from '../utils/map-props.js';
import { withInstall } from '../utils/withInstall.js';
import './style/css.js';
import './type.js';
import '@babel/runtime/helpers/defineProperty';
import 'vue';
import '../config.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/typeof';
import 'tdesign-icons-vue-next';
import '../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../config-provider/config-receiver.js';
import '../config-provider/zh_CN_config.js';
import '../config-provider/type.js';
import '../utils/mixins.js';
import '../utils/classnames.js';
import './props.js';
import '../utils/event.js';
import '../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isString';
import 'lodash/isFunction';
import 'lodash/isObject';
import 'lodash/kebabCase';

var Addon = withInstall(_Addon);
var Input = withInstall(mapProps([{
  name: "value",
  event: ["change"],
  alias: ["modelValue"]
}])(_Input));
var InputGroup = withInstall(_InputGroup);

export { Addon, Input, InputGroup, Input as default };
