// tdesign-vue-next/lib/utils/classnames.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../config.js');

var SIZE_CLASSNAMES = {
  small: "".concat(config.prefix, "-size-s"),
  medium: "".concat(config.prefix, "-size-m"),
  large: "".concat(config.prefix, "-size-l"),
  "default": "",
  xs: "".concat(config.prefix, "-size-xs"),
  xl: "".concat(config.prefix, "-size-xl"),
  block: "".concat(config.prefix, "-size-full-width")
};
var STATUS_CLASSNAMES = {
  loading: "".concat(config.prefix, "-is-loading"),
  loadMore: "".concat(config.prefix, "-is-load-more"),
  disabled: "".concat(config.prefix, "-is-disabled"),
  focused: "".concat(config.prefix, "-is-focused"),
  success: "".concat(config.prefix, "-is-success"),
  error: "".concat(config.prefix, "-is-error"),
  warning: "".concat(config.prefix, "-is-warning"),
  selected: "".concat(config.prefix, "-is-selected"),
  active: "".concat(config.prefix, "-is-active"),
  checked: "".concat(config.prefix, "-is-checked"),
  current: "".concat(config.prefix, "-is-current"),
  hidden: "".concat(config.prefix, "-is-hidden"),
  visible: "".concat(config.prefix, "-is-visible"),
  expanded: "".concat(config.prefix, "-is-expanded"),
  indeterminate: "".concat(config.prefix, "-is-indeterminate")
};
var ClASSNAMES = {
  SIZE: SIZE_CLASSNAMES,
  STATUS: STATUS_CLASSNAMES
};

exports.SIZE_CLASSNAMES = SIZE_CLASSNAMES;
exports.STATUS_CLASSNAMES = STATUS_CLASSNAMES;
exports["default"] = ClASSNAMES;
