// tdesign-vue-next/lib/upload/dragger.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var loading_index = require('../loading/index.js');
var button_index = require('../button/index.js');
var upload_util = require('./util.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/slicedToArray');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../utils/classnames.js');
require('../utils/dom.js');
require('../_chunks/dep-eb6b0f94.js');
require('lodash/isString');
require('../utils/easing.js');
require('../utils/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isObject');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../utils/withInstall.js');
require('../loading/style');
require('../loading/plugin.js');
require('../button/button.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../button/props.js');
require('../utils/ripple.js');
require('../utils/set-style.js');
require('../button/style');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var name = "".concat(config.prefix, "-upload-dragger");
var Dragger = vue.defineComponent({
  name: name,
  components: {
    TButton: button_index.Button,
    TLoading: loading_index.Loading,
    CheckCircleFilledIcon: tdesignIconsVueNext.CheckCircleFilledIcon,
    ErrorCircleFilledIcon: tdesignIconsVueNext.ErrorCircleFilledIcon
  },
  props: {
    file: {
      type: Object
    },
    loadingFile: {
      type: Object
    },
    display: {
      type: String,
      validator: function validator(val) {
        return ["file", "file-input", "image", "custom"].includes(val);
      }
    },
    cancel: Function,
    trigger: Function,
    remove: Function,
    upload: Function,
    autoUpload: Boolean
  },
  emits: ["change", "dragleave", "dragenter"],
  data: function data() {
    return {
      target: null,
      dragActive: false
    };
  },
  computed: {
    isImage: function isImage() {
      return this.display === "image";
    },
    imageUrl: function imageUrl() {
      return this.loadingFile && this.loadingFile.url || this.file && this.file.url;
    },
    percent: function percent() {
      return this.loadingFile && this.loadingFile.percent;
    },
    inputName: function inputName() {
      return this.loadingFile && this.loadingFile.name || this.file && this.file.name;
    },
    classes: function classes() {
      return ["".concat(upload_util.UPLOAD_NAME, "__dragger"), _defineProperty__default["default"]({}, "".concat(upload_util.UPLOAD_NAME, "__dragger-center"), !this.loadingFile && !this.file), _defineProperty__default["default"]({}, "".concat(upload_util.UPLOAD_NAME, "__dragger-error"), this.loadingFile && this.loadingFile.status === "fail")];
    },
    size: function size() {
      return this.loadingFile && this.loadingFile.size || this.file && this.file.size;
    },
    showResultOperate: function showResultOperate() {
      var _this$file, _this$loadingFile;

      return Boolean(!this.loadingFile && ((_this$file = this.file) === null || _this$file === void 0 ? void 0 : _this$file.name)) || ["success", "fail"].includes((_this$loadingFile = this.loadingFile) === null || _this$loadingFile === void 0 ? void 0 : _this$loadingFile.status);
    }
  },
  methods: {
    handleDrop: function handleDrop(event) {
      event.preventDefault();
      this.$emit("change", event.dataTransfer.files);
      this.$emit("dragleave", event);
      this.dragActive = false;
    },
    handleDragenter: function handleDragenter(event) {
      this.target = event.target;
      event.preventDefault();
      this.$emit("dragenter", event);
      this.dragActive = true;
    },
    handleDragleave: function handleDragleave(event) {
      if (this.target !== event.target) return;
      event.preventDefault();
      this.$emit("dragleave", event);
      this.dragActive = false;
    },
    handleDragover: function handleDragover(event) {
      event.preventDefault();
    },
    renderDefaultDragElement: function renderDefaultDragElement() {
      var unActiveElement = vue.createVNode("div", null, [vue.createVNode("span", {
        "class": "".concat(config.prefix, "-upload--highlight")
      }, ["\u70B9\u51FB\u4E0A\u4F20"]), vue.createVNode("span", null, ["\xA0\xA0/\xA0\xA0\u62D6\u62FD\u5230\u6B64\u533A\u57DF"])]);

      var activeElement = vue.createVNode("div", null, ["\u91CA\u653E\u9F20\u6807"]);

      return this.dragActive ? activeElement : unActiveElement;
    },
    renderImage: function renderImage() {
      return vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-img-wrap")
      }, [this.imageUrl && vue.createVNode("img", {
        "src": this.imageUrl
      }, null)]);
    },
    renderUploading: function renderUploading() {
      if (this.loadingFile.status === "fail") {
        return vue.createVNode(tdesignIconsVueNext.ErrorCircleFilledIcon, null, null);
      }

      if (this.loadingFile.status === "progress") {
        return vue.createVNode("div", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__single-progress")
        }, [vue.createVNode(loading_index.Loading, null, null), vue.createVNode("span", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__single-percent")
        }, [Math.min(this.loadingFile.percent, 99), "%"])]);
      }
    },
    reupload: function reupload(e) {
      this.remove(e);
      this.trigger(e);
    },
    renderProgress: function renderProgress() {
      var _this$loadingFile2,
          _this$loadingFile3,
          _this = this;

      return vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-progress")
      }, [this.isImage && this.renderImage(), vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-progress-info")
      }, [vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-text")
      }, [vue.createVNode("span", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__single-name")
      }, [upload_util.abridgeName(this.inputName)]), this.loadingFile && this.renderUploading(), !this.loadingFile && !!this.file && vue.createVNode(tdesignIconsVueNext.CheckCircleFilledIcon, null, null)]), vue.createVNode("small", {
        "class": "".concat(config.prefix, "-size-s")
      }, ["\u6587\u4EF6\u5927\u5C0F\uFF1A", upload_util.returnFileSize(this.size)]), vue.createVNode("small", {
        "class": "".concat(config.prefix, "-size-s")
      }, ["\u4E0A\u4F20\u65E5\u671F\uFF1A", upload_util.getCurrentDate()]), vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-btns")
      }, [["progress", "waiting"].includes((_this$loadingFile2 = this.loadingFile) === null || _this$loadingFile2 === void 0 ? void 0 : _this$loadingFile2.status) && vue.createVNode(button_index.Button, {
        "theme": "primary",
        "variant": "text",
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-progress-cancel"),
        "onClick": this.cancel
      }, {
        "default": function _default() {
          return ["\u53D6\u6D88\u4E0A\u4F20"];
        }
      }), !this.autoUpload && ((_this$loadingFile3 = this.loadingFile) === null || _this$loadingFile3 === void 0 ? void 0 : _this$loadingFile3.status) === "waiting" && vue.createVNode(button_index.Button, {
        "theme": "primary",
        "variant": "text",
        "onClick": function onClick(e) {
          return _this.upload(_objectSpread({}, _this.loadingFile), e);
        }
      }, {
        "default": function _default() {
          return ["\u5F00\u59CB\u4E0A\u4F20"];
        }
      })]), this.showResultOperate && vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-btns")
      }, [vue.createVNode(vue.resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "text",
        "class": "".concat(upload_util.UPLOAD_NAME, "__dragger-progress-cancel"),
        "onClick": this.reupload
      }, {
        "default": function _default() {
          return ["\u91CD\u65B0\u4E0A\u4F20"];
        }
      }), vue.createVNode(vue.resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "text",
        "onClick": this.remove
      }, {
        "default": function _default() {
          return ["\u5220\u9664"];
        }
      })])])]);
    }
  },
  render: function render() {
    var content = null;

    if ((this.loadingFile || this.file) && this.display !== "custom") {
      content = this.renderProgress();
    } else {
      content = vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__trigger"),
        "onClick": this.trigger
      }, [this.$slots["default"] && this.$slots["default"](null) || this.renderDefaultDragElement()]);
    }

    return vue.createVNode("div", {
      "class": this.classes,
      "onDrop": this.handleDrop,
      "onDragenter": this.handleDragenter,
      "onDragover": this.handleDragover,
      "onDragleave": this.handleDragleave
    }, [content]);
  }
});

exports["default"] = Dragger;
