// tdesign-vue-next/lib/upload/flow-list.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var button_index = require('../button/index.js');
var loading_index = require('../loading/index.js');
var upload_util = require('./util.js');
var upload_props = require('./props.js');
require('../button/button.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../utils/classnames.js');
require('../button/props.js');
require('../utils/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isString');
require('lodash/isFunction');
require('lodash/isObject');
require('../utils/ripple.js');
require('../utils/set-style.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/slicedToArray');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../utils/dom.js');
require('../_chunks/dep-eb6b0f94.js');
require('../utils/easing.js');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../utils/withInstall.js');
require('../loading/style');
require('../loading/plugin.js');
require('../button/style');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var FlowList = vue.defineComponent({
  name: "TUploadFlowList",
  components: {
    TButton: button_index.Button,
    TLoading: loading_index.Loading,
    TimeFilledIcon: tdesignIconsVueNext.TimeFilledIcon,
    CheckCircleFilledIcon: tdesignIconsVueNext.CheckCircleFilledIcon,
    ErrorCircleFilledIcon: tdesignIconsVueNext.ErrorCircleFilledIcon,
    DeleteIcon: tdesignIconsVueNext.DeleteIcon,
    BrowseIcon: tdesignIconsVueNext.BrowseIcon
  },
  props: {
    showUploadProgress: upload_props["default"].showUploadProgress,
    files: Array,
    toUploadFiles: Array,
    placeholder: String,
    autoUpload: Boolean,
    remove: Function,
    upload: Function,
    cancel: Function,
    display: {
      type: String,
      validator: function validator(val) {
        return ["file-flow", "image-flow"].includes(val);
      }
    }
  },
  emits: ["dragleave", "change", "dragenter", "imgPreview"],
  data: function data() {
    return {
      dragActive: false,
      target: null
    };
  },
  computed: {
    showInitial: function showInitial() {
      var isWatingEmpty = !this.waitingUploadFiles || !this.waitingUploadFiles.length;
      return (!this.files || !this.files.length) && isWatingEmpty;
    },
    waitingUploadFiles: function waitingUploadFiles() {
      var _this = this;

      var list = [];
      this.toUploadFiles.forEach(function (item) {
        var r = _this.files.filter(function (t) {
          return t.name === item.name;
        });

        if (!r.length) {
          list.push(item);
        }
      });
      return list;
    },
    listFiles: function listFiles() {
      if (!this.files || !this.files.length) return this.toUploadFiles;
      return this.files.concat(this.waitingUploadFiles);
    },
    failedList: function failedList() {
      return this.toUploadFiles.filter(function (file) {
        return file.status === "fail";
      });
    },
    processList: function processList() {
      return this.toUploadFiles.filter(function (file) {
        return file.status === "progress";
      });
    },
    isUploading: function isUploading() {
      return !!this.processList.length;
    },
    allowUpload: function allowUpload() {
      return Boolean(this.waitingUploadFiles && this.waitingUploadFiles.length) && !this.isUploading;
    },
    uploadText: function uploadText() {
      if (this.isUploading) return "\u4E0A\u4F20\u4E2D...";
      return this.failedList && this.failedList.length ? "\u91CD\u65B0\u4E0A\u4F20" : "\u5F00\u59CB\u4E0A\u4F20";
    }
  },
  methods: {
    renderStatus: function renderStatus(file) {
      var status = null;

      switch (file.status) {
        case "success":
          status = vue.createVNode("div", {
            "class": "".concat(upload_util.UPLOAD_NAME, "__flow-status")
          }, [vue.createVNode(tdesignIconsVueNext.CheckCircleFilledIcon, null, null), vue.createVNode("span", null, ["\u4E0A\u4F20\u6210\u529F"])]);
          break;

        case "fail":
          status = vue.createVNode("div", {
            "class": "".concat(upload_util.UPLOAD_NAME, "__flow-status")
          }, [vue.createVNode(tdesignIconsVueNext.ErrorCircleFilledIcon, null, null), vue.createVNode("span", null, ["\u4E0A\u4F20\u5931\u8D25"])]);
          break;

        case "progress":
          this.showUploadProgress && (status = vue.createVNode("div", {
            "class": "".concat(upload_util.UPLOAD_NAME, "__flow-status")
          }, [vue.createVNode(loading_index.Loading, null, null), vue.createVNode("span", null, ["\u4E0A\u4F20\u4E2D ", Math.min(file.percent, 99), "%"])]));
          break;

        case "waiting":
          status = vue.createVNode("div", {
            "class": "".concat(upload_util.UPLOAD_NAME, "__flow-status")
          }, [vue.createVNode(tdesignIconsVueNext.TimeFilledIcon, null, null), vue.createVNode("span", null, ["\u5F85\u4E0A\u4F20"])]);
          break;
      }

      return status;
    },
    handleDrop: function handleDrop(event) {
      event.preventDefault();
      this.$emit("change", event.dataTransfer.files);
      this.$emit("dragleave", event);
      this.dragActive = false;
    },
    handleDragenter: function handleDragenter(event) {
      this.target = event.target;
      event.preventDefault();
      this.$emit("dragenter", event);
      this.dragActive = true;
    },
    handleDragleave: function handleDragleave(event) {
      if (this.target !== event.target) return;
      event.preventDefault();
      this.$emit("dragleave", event);
      this.dragActive = false;
    },
    handleDragover: function handleDragover(event) {
      event.preventDefault();
    },
    onViewClick: function onViewClick(event, file) {
      this.$emit("imgPreview", event, file);
    },
    renderDrager: function renderDrager() {
      return vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__flow-empty"),
        "onDrop": this.handleDrop,
        "onDragenter": this.handleDragenter,
        "onDragover": this.handleDragover,
        "onDragleave": this.handleDragleave
      }, [this.dragActive ? "\u91CA\u653E\u9F20\u6807" : "\u70B9\u51FB\u4E0A\u65B9\u201C\u9009\u62E9\u6587\u4EF6\u201D\u6216\u5C06\u6587\u4EF6\u62D6\u62FD\u5230\u6B64\u533A\u57DF"]);
    },
    renderFileList: function renderFileList() {
      var _this2 = this;

      return vue.createVNode("table", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__flow-table")
      }, [vue.createVNode("tr", null, [vue.createVNode("th", null, ["\u6587\u4EF6\u540D"]), vue.createVNode("th", null, ["\u5927\u5C0F"]), vue.createVNode("th", null, ["\u72B6\u6001"]), vue.createVNode("th", null, ["\u64CD\u4F5C"])]), this.showInitial && vue.createVNode("tr", null, [vue.createVNode("td", {
        "colspan": 4
      }, [this.renderDrager()])]), this.listFiles.map(function (file, index) {
        return vue.createVNode("tr", null, [vue.createVNode("td", null, [upload_util.abridgeName(file.name, 7, 10)]), vue.createVNode("td", null, [upload_util.returnFileSize(file.size)]), vue.createVNode("td", null, [_this2.renderStatus(file)]), vue.createVNode("td", null, [vue.createVNode("span", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__flow-button"),
          "onClick": function onClick(e) {
            return _this2.remove({
              e: e,
              index: index,
              file: file
            });
          }
        }, ["\u5220\u9664"])])]);
      })]);
    },
    renderImgList: function renderImgList() {
      var _this3 = this;

      return vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__flow-card-area")
      }, [this.showInitial && this.renderDrager(), !!this.listFiles.length && vue.createVNode("ul", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card clearfix")
      }, [this.listFiles.map(function (file, index) {
        return vue.createVNode("li", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-item")
        }, [vue.createVNode("div", {
          "class": ["".concat(upload_util.UPLOAD_NAME, "__card-content"), _defineProperty__default["default"]({}, "".concat(config.prefix, "-is-bordered"), file.status !== "waiting")]
        }, [file.status === "fail" && vue.createVNode("div", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-status-wrap")
        }, [vue.createVNode(tdesignIconsVueNext.ErrorCircleFilledIcon, null, null), vue.createVNode("p", null, ["\u4E0A\u4F20\u5931\u8D25"])]), file.status === "progress" && vue.createVNode("div", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-status-wrap")
        }, [vue.createVNode(loading_index.Loading, null, null), vue.createVNode("p", null, ["\u4E0A\u4F20\u4E2D ", Math.min(file.percent, 99)])]), (["waiting", "success"].includes(file.status) || !file.status && file.url) && vue.createVNode("img", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-image"),
          "src": file.url || "//tdesign.gtimg.com/tdesign-default-img.png"
        }, null), vue.createVNode("div", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask")
        }, [file.url && vue.createVNode("span", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask-item")
        }, [vue.createVNode(tdesignIconsVueNext.BrowseIcon, {
          "onClick": function onClick(_ref2) {
            var e = _ref2.e;
            return _this3.onViewClick(e, file);
          }
        }, null), vue.createVNode("span", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask-item-divider")
        }, null)]), vue.createVNode("span", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask-item"),
          "onClick": function onClick(e) {
            return _this3.remove({
              e: e,
              index: index,
              file: file
            });
          }
        }, [vue.createVNode(tdesignIconsVueNext.DeleteIcon, null, null)])])]), vue.createVNode("p", {
          "class": "".concat(upload_util.UPLOAD_NAME, "__card-name")
        }, [upload_util.abridgeName(file.name)])]);
      })])]);
    }
  },
  render: function render() {
    var _this4 = this;

    return vue.createVNode("div", {
      "class": ["".concat(upload_util.UPLOAD_NAME, "__flow"), "".concat(upload_util.UPLOAD_NAME, "__flow-").concat(this.display)]
    }, [vue.createVNode("div", {
      "class": "".concat(upload_util.UPLOAD_NAME, "__flow-op")
    }, [this.$slots["default"] && this.$slots["default"](null), vue.createVNode("small", {
      "class": "".concat(config.prefix, "-size-s ").concat(upload_util.UPLOAD_NAME, "__flow-placeholder")
    }, [this.placeholder])]), this.display === "file-flow" && this.renderFileList(), this.display === "image-flow" && this.renderImgList(), vue.createVNode("div", {
      "class": "".concat(upload_util.UPLOAD_NAME, "__flow-bottom")
    }, [vue.createVNode(button_index.Button, {
      "theme": "default",
      "onClick": this.cancel
    }, {
      "default": function _default() {
        return ["\u53D6\u6D88"];
      }
    }), vue.createVNode(button_index.Button, {
      "disabled": !this.allowUpload,
      "theme": "primary",
      "onClick": function onClick(e) {
        return _this4.upload(_this4.waitingUploadFiles, e);
      }
    }, {
      "default": function _default() {
        return [_this4.uploadText];
      }
    })])]);
  }
});

exports["default"] = FlowList;
