// tdesign-vue-next/lib/upload/image.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var loading_index = require('../loading/index.js');
var upload_util = require('./util.js');
var upload_props = require('./props.js');
var utils_classnames = require('../utils/classnames.js');
var config = require('../config.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/slicedToArray');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../utils/dom.js');
require('../_chunks/dep-eb6b0f94.js');
require('lodash/isString');
require('../utils/easing.js');
require('../utils/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isObject');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../utils/withInstall.js');
require('../loading/style');
require('../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var ImageCard = vue.defineComponent({
  name: "TImageUpload",
  components: {
    AddIcon: tdesignIconsVueNext.AddIcon,
    DeleteIcon: tdesignIconsVueNext.DeleteIcon,
    UploadIcon: tdesignIconsVueNext.UploadIcon,
    BrowseIcon: tdesignIconsVueNext.BrowseIcon,
    TLoading: loading_index.Loading
  },
  props: {
    showUploadProgress: upload_props["default"].showUploadProgress,
    files: {
      type: Array
    },
    loadingFile: {
      type: Object
    },
    trigger: {
      type: Function
    },
    remove: {
      type: Function
    },
    multiple: Boolean,
    max: Number,
    disabled: Boolean
  },
  emits: ["img-preview"],
  computed: {
    showTrigger: function showTrigger() {
      if (this.multiple) {
        return !this.max || this.files && this.files.length < this.max;
      }

      return !(this.files && this.files[0]);
    }
  },
  methods: {
    onMaskClick: function onMaskClick(e) {
      !this.showTrigger && this.trigger(e);
    },
    onViewClick: function onViewClick(e, file) {
      this.$emit("img-preview", e, file);
    }
  },
  render: function render() {
    var _this = this;

    return vue.createVNode("ul", {
      "class": "".concat(upload_util.UPLOAD_NAME, "__card")
    }, [this.files && this.files.map(function (file, index) {
      return vue.createVNode("li", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-item ").concat(config.prefix, "-is--background")
      }, [vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-content ").concat(upload_util.UPLOAD_NAME, "__card-box")
      }, [vue.createVNode("img", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-image"),
        "src": file.url
      }, null), vue.createVNode("div", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask"),
        "onClick": _this.onMaskClick
      }, [vue.createVNode("span", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask-item"),
        "onClick": function onClick(e) {
          return e.stopPropagation();
        }
      }, [vue.createVNode(tdesignIconsVueNext.BrowseIcon, {
        "onClick": function onClick(_ref) {
          var e = _ref.e;
          return _this.onViewClick(e, file);
        }
      }, null)]), vue.createVNode("span", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask-item-divider")
      }, null), vue.createVNode("span", {
        "class": "".concat(upload_util.UPLOAD_NAME, "__card-mask-item"),
        "onClick": function onClick(e) {
          return e.stopPropagation();
        }
      }, [vue.createVNode(tdesignIconsVueNext.DeleteIcon, {
        "onClick": function onClick(_ref2) {
          var e = _ref2.e;
          return _this.remove({
            e: e,
            file: file,
            index: index
          });
        }
      }, null)])])])]);
    }), this.showTrigger && vue.createVNode("li", {
      "class": ["".concat(upload_util.UPLOAD_NAME, "__card-item ").concat(config.prefix, "-is--background"), _defineProperty__default["default"]({}, utils_classnames["default"].STATUS.disabled, this.disabled)],
      "onClick": this.trigger
    }, [this.showUploadProgress && this.loadingFile && this.loadingFile.status === "progress" ? vue.createVNode("div", {
      "class": "".concat(upload_util.UPLOAD_NAME, "__card-container ").concat(upload_util.UPLOAD_NAME, "__card-box")
    }, [vue.createVNode(loading_index.Loading, null, null), vue.createVNode("p", null, ["\u4E0A\u4F20\u4E2D ", Math.min(this.loadingFile.percent, 99), "%"])]) : vue.createVNode("div", {
      "class": "".concat(upload_util.UPLOAD_NAME, "__card-container ").concat(upload_util.UPLOAD_NAME, "__card-box")
    }, [vue.createVNode(tdesignIconsVueNext.AddIcon, null, null), vue.createVNode("p", {
      "class": "".concat(config.prefix, "-size-s")
    }, ["\u70B9\u51FB\u4E0A\u4F20\u56FE\u7247"])])])]);
  }
});

exports["default"] = ImageCard;
