// tdesign-vue-next/esm/drawer/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

var props = {
  attach: {
    type: [String, Function],
    "default": ""
  },
  body: {
    type: [String, Function]
  },
  cancelBtn: {
    type: [String, Object, Function],
    "default": ""
  },
  closeBtn: {
    type: [String, Boolean, Function],
    "default": void 0
  },
  closeOnEscKeydown: {
    type: Boolean,
    "default": true
  },
  closeOnOverlayClick: {
    type: Boolean,
    "default": true
  },
  confirmBtn: {
    type: [String, Object, Function],
    "default": ""
  },
  "default": {
    type: [String, Function]
  },
  destroyOnClose: Boolean,
  footer: {
    type: [Boolean, Function],
    "default": true
  },
  header: {
    type: [String, Boolean, Function],
    "default": void 0
  },
  mode: {
    type: String,
    "default": "overlay",
    validator: function validator(val) {
      return ["overlay", "push"].includes(val);
    }
  },
  placement: {
    type: String,
    "default": "right",
    validator: function validator(val) {
      return ["left", "right", "top", "bottom"].includes(val);
    }
  },
  preventScrollThrough: {
    type: Boolean,
    "default": true
  },
  showInAttachedElement: Boolean,
  showOverlay: {
    type: Boolean,
    "default": true
  },
  size: {
    type: String,
    "default": "small"
  },
  sizeDraggable: Boolean,
  visible: Boolean,
  zIndex: {
    type: Number
  },
  onCancel: Function,
  onClose: Function,
  onCloseBtnClick: Function,
  onConfirm: Function,
  onEscKeydown: Function,
  onOverlayClick: Function
};

export { props as default };
