// tdesign-vue-next/esm/tree/constants.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import { prefix } from '../config.js';

var TREE_NAME = "".concat(prefix, "-tree");
var TREE_NODE_NAME = "".concat(prefix, "-tree-node");
var FX = {
  treeNode: "".concat(prefix, "-tree-toggle")
};
var tree = "".concat(prefix, "-tree");
var CLASS_NAMES = {
  icon: "".concat(prefix, "-icon"),
  folderIcon: "".concat(prefix, "-folder-icon"),
  actived: "".concat(prefix, "-is-active"),
  disabled: "".concat(prefix, "-is-disabled"),
  treeIconRight: "".concat(prefix, "-icon-arrow-right"),
  treeIconDown: "".concat(prefix, "-icon-arrow-down"),
  tree: tree,
  treeTransition: "".concat(tree, "--transition"),
  treeBlockNode: "".concat(tree, "--block-node"),
  treeEmpty: "".concat(tree, "__empty"),
  treeList: "".concat(tree, "__list"),
  treeNode: "".concat(tree, "__item"),
  treeNodeOpen: "".concat(tree, "__item--open"),
  treeHoverable: "".concat(tree, "--hoverable"),
  treeCheckable: "".concat(tree, "--checkable"),
  treeLabel: "".concat(tree, "__label"),
  treeLabelStrictly: "".concat(tree, "__label--strictly"),
  treeIcon: "".concat(tree, "__icon"),
  treeIconDefault: "".concat(tree, "__icon--default"),
  treeSpace: "".concat(tree, "__space"),
  treeOperations: "".concat(tree, "__operations"),
  treeNodeVisible: "".concat(tree, "__item--visible"),
  treeNodeHidden: "".concat(tree, "__item--hidden"),
  treeNodeEnter: "".concat(tree, "__item--enter-active"),
  treeNodeLeave: "".concat(tree, "__item--leave-active"),
  line: "".concat(tree, "__line"),
  lineIsLeaf: "".concat(tree, "__line--leaf"),
  lineIsFirst: "".concat(tree, "__line--first")
};

export { CLASS_NAMES, FX, TREE_NAME, TREE_NODE_NAME };
