// tdesign-vue-next/esm/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import { c as components } from './_chunks/dep-741293b1.js';
export { Button } from './button/index.js';
export { Divider } from './divider/index.js';
export { Icon } from './icon/index.js';
export { Col, Row } from './grid/index.js';
export { Aside, Content, Footer, Header, Layout } from './layout/index.js';
export { Affix } from './affix/index.js';
export { Anchor, AnchorItem, AnchorTarget } from './anchor/index.js';
export { Breadcrumb, BreadcrumbItem } from './breadcrumb/index.js';
export { Dropdown, DropdownItem, DropdownMenu } from './dropdown/index.js';
export { HeadMenu, Menu, MenuGroup, MenuItem, Submenu } from './menu/index.js';
export { Pagination } from './pagination/index.js';
export { StepItem, Steps } from './steps/index.js';
export { TabPanel, Tabs } from './tabs/index.js';
export { Cascader } from './cascader/index.js';
export { Checkbox, CheckboxGroup } from './checkbox/index.js';
export { DatePicker } from './date-picker/index.js';
export { Form, FormItem } from './form/index.js';
export { Addon, Input, InputGroup } from './input/index.js';
export { InputNumber } from './input-number/index.js';
export { Radio, RadioButton, RadioGroup } from './radio/index.js';
export { Option, OptionGroup, Select } from './select/index.js';
export { Slider, SliderButton } from './slider/index.js';
export { Switch } from './switch/index.js';
export { TagInput } from './tag-input/index.js';
export { Textarea } from './textarea/index.js';
export { Transfer } from './transfer/index.js';
export { TimePicker, TimeRangePicker } from './time-picker/index.js';
export { TreeSelect } from './tree-select/index.js';
export { Avatar, AvatarGroup } from './avatar/index.js';
export { Badge } from './badge/index.js';
export { Calendar } from './calendar/index.js';
export { Comment } from './comment/index.js';
export { List, ListItem, ListItemMeta } from './list/index.js';
export { Progress } from './progress/index.js';
export { Skeleton } from './skeleton/index.js';
export { Swiper, SwiperItem } from './swiper/index.js';
export { BaseTable, EnhancedTable, PrimaryTable } from './table/index.js';
export { CheckTag, Tag } from './tag/index.js';
export { Tooltip } from './tooltip/index.js';
export { Tree } from './tree/index.js';
export { Alert } from './alert/index.js';
export { default as DialogPlugin } from './dialog/plugin.js';
export { Dialog } from './dialog/index.js';
export { Drawer } from './drawer/index.js';
export { default as LoadingPlugin } from './loading/plugin.js';
export { Loading } from './loading/index.js';
export { default as MessagePlugin } from './message/plugin.js';
export { Message } from './message/index.js';
export { default as NotifyPlugin } from './notification/plugin.js';
export { Notification } from './notification/index.js';
export { Popconfirm } from './popconfirm/index.js';
export { Popup } from './popup/index.js';
export { Upload } from './upload/index.js';
export { useReceiver } from './config-provider/useReceiver.js';
export { ConfigProvider } from './config-provider/index.js';
import './button/button.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/defineProperty';
import 'vue';
import './config.js';
import './utils/classnames.js';
import './button/props.js';
import './utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isString';
import 'lodash/isFunction';
import 'lodash/isObject';
import './utils/ripple.js';
import './utils/set-style.js';
import './loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import './loading/icon/gradient.js';
import './_common/js/loading/circle-adapter.js';
import './_common/js/utils/set-style.js';
import './_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import './utils/dom.js';
import './_chunks/dep-b66bfe36.js';
import './utils/easing.js';
import './utils/transfer-dom.js';
import './loading/props.js';
import './utils/withInstall.js';
import './loading/style';
import './loading/type.js';
import './button/style';
import './button/type.js';
import './divider/divider.js';
import './divider/props.js';
import './divider/style';
import './divider/type.js';
import 'tdesign-icons-vue-next';
import './grid/row.js';
import './grid/row-props.js';
import './grid/common.js';
import './utils/responsive.js';
import './hooks/event.js';
import './utils/helper.js';
import 'lodash/camelCase';
import './grid/col.js';
import './grid/col-props.js';
import './grid/style';
import './grid/type.js';
import './layout/layout.js';
import './layout/header.js';
import './layout/header-props.js';
import './layout/footer.js';
import './layout/footer-props.js';
import './layout/aside.js';
import './layout/aside-props.js';
import './layout/content.js';
import './layout/style';
import './layout/type.js';
import './affix/affix.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import './affix/props.js';
import './affix/style';
import './affix/type.js';
import './anchor/anchor.js';
import './anchor/utils.js';
import './anchor/props.js';
import './anchor/constant.js';
import './utils/event.js';
import './anchor/anchor-item.js';
import './anchor/anchor-item-props.js';
import './anchor/anchor-target.js';
import './utils/clipboard.js';
import './anchor/anchor-target-props.js';
import '@babel/runtime/helpers/typeof';
import './message/messageList.js';
import './message/const.js';
import './message/message.js';
import './message/props.js';
import './popup/popup.js';
import '@popperjs/core';
import './popup/props.js';
import './utils/map-props.js';
import 'lodash/kebabCase';
import './popup/style';
import './popup/type.js';
import './anchor/style';
import './anchor/type.js';
import './breadcrumb/breadcrumb.js';
import './breadcrumb/props.js';
import './breadcrumb/breadcrumb-item.js';
import './breadcrumb/breadcrumb-item-props.js';
import './tooltip/tooltip.js';
import './tooltip/props.js';
import './tooltip/style';
import './tooltip/type.js';
import './breadcrumb/style';
import './breadcrumb/type.js';
import './dropdown/dropdown-item.js';
import './dropdown/dropdown-item-props.js';
import './dropdown/dropdown.js';
import './dropdown/dropdown-menu.js';
import './dropdown/props.js';
import './dropdown/style';
import './dropdown/type.js';
import './menu/menu.js';
import './menu/props.js';
import './menu/v-menu.js';
import '@babel/runtime/helpers/classCallCheck';
import '@babel/runtime/helpers/createClass';
import './_common/js/log/log.js';
import './menu/head-menu.js';
import './menu/head-menu-props.js';
import './tabs/tabs.js';
import './tabs/tab-panel.js';
import './tabs/tab-panel-props.js';
import './tabs/tab-nav.js';
import 'lodash/debounce';
import './tabs/tab-nav-item.js';
import './tabs/props.js';
import './tabs/style';
import './tabs/type.js';
import './menu/submenu.js';
import './menu/submenu-props.js';
import './common-components/fake-arrow.js';
import './menu/menu-item.js';
import './menu/menu-item-props.js';
import './menu/menu-group.js';
import './menu/menu-group-props.js';
import './menu/style';
import './menu/type.js';
import './pagination/pagination.js';
import './utils/mixins.js';
import './config-provider/config-receiver.js';
import './config-provider/zh_CN_config.js';
import './config-provider/type.js';
import './pagination/props.js';
import './input-number/input-number.js';
import './input-number/props.js';
import './input-number/style';
import './input-number/type.js';
import './select/select.js';
import 'lodash/get';
import 'lodash/set';
import './select/option.js';
import './select/option-props.js';
import './checkbox/checkbox.js';
import './checkbox/props.js';
import './checkbox/group.js';
import 'lodash/intersection';
import './checkbox/checkbox-group-props.js';
import './checkbox/style';
import './checkbox/type.js';
import './select/props.js';
import './input/addon.js';
import './input/input.js';
import './input/props.js';
import './input/input-group.js';
import './input/style';
import './input/type.js';
import './tag/tag.js';
import './tag/props.js';
import './tag/check-tag.js';
import './tag/check-tag-props.js';
import './tag/style';
import './tag/type.js';
import './select/optionGroup.js';
import './select/option-group-props.js';
import './select/style';
import './select/type.js';
import './pagination/style';
import './pagination/type.js';
import './steps/steps.js';
import './steps/props.js';
import './steps/step-item.js';
import './steps/step-item-props.js';
import './steps/style';
import './steps/type.js';
import './cascader/cascader.js';
import 'lodash/isEqual';
import './_common/js/tree/tree-store.js';
import 'lodash/difference';
import 'lodash/isPlainObject';
import './_common/js/tree/tree-node.js';
import 'lodash/uniqueId';
import './_common/js/tree/tree-node-model.js';
import 'lodash/pick';
import './cascader/utils/helper.js';
import './cascader/utils/cascader.js';
import './cascader/components/Panel.js';
import './cascader/utils/panel.js';
import 'lodash/isNumber';
import './cascader/components/Item.js';
import './cascader/utils/item.js';
import './cascader/props.js';
import './cascader/components/InputContent.js';
import './cascader/utils/inputContent.js';
import './cascader/interface.js';
import './cascader/type.js';
import './cascader/style';
import './date-picker/date-picker.js';
import 'dayjs';
import './date-picker/props.js';
import './date-picker/calendar-presets.js';
import './date-picker/panel/date.js';
import './date-picker/basic/header.js';
import './date-picker/basic/table.js';
import './date-picker/basic/cell.js';
import './_common/js/date-picker/utils.js';
import '@babel/runtime/helpers/extends';
import 'lodash/chunk';
import './date-picker/panel/date-range.js';
import './time-picker/panel/index.js';
import './_chunks/dep-ff8379e8.js';
import './time-picker/panel/props.js';
import './time-picker/props.js';
import './time-picker/constant.js';
import './date-picker/style';
import './date-picker/type.js';
import './form/form.js';
import './form/props.js';
import './form/const.js';
import './form/form-item.js';
import 'lodash/cloneDeep';
import 'lodash/isNil';
import './form/form-model.js';
import './form/form-item-props.js';
import './form/style';
import './form/type.js';
import './radio/radio.js';
import './radio/props.js';
import './radio/group.js';
import './radio/radio-group-props.js';
import './radio/radio-button.js';
import './radio/style';
import './radio/type.js';
import './slider/slider.js';
import './slider/props.js';
import './slider/slider-mark.js';
import './slider/slider-button.js';
import './slider/style';
import './slider/type.js';
import './switch/switch.js';
import './switch/props.js';
import './switch/style';
import './switch/type.js';
import './tag-input/tag-input.js';
import './tag-input/props.js';
import './tag-input/useTagScroll.js';
import './tag-input/useTagList.js';
import './hooks/useDefault.js';
import './tag-input/useHover.js';
import './tag-input/style';
import './tag-input/type.js';
import './textarea/textarea.js';
import './textarea/props.js';
import './textarea/calcTextareaHeight.js';
import './textarea/style';
import './textarea/type.js';
import './transfer/transfer.js';
import './transfer/components/transfer-list.js';
import './transfer/utils.js';
import './transfer/components/transfer-search.js';
import './transfer/components/transfer-operations.js';
import './transfer/props.js';
import './transfer/style';
import './time-picker/time-picker.js';
import './time-picker/input-items.js';
import './time-picker/time-range-picker.js';
import './time-picker/time-range-picker-props.js';
import './time-picker/style';
import './time-picker/type.js';
import './tree-select/tree-select.js';
import 'lodash/isArray';
import 'lodash/isBoolean';
import './tree-select/props.js';
import './tree/td-tree.js';
import 'lodash/upperFirst';
import './tree/tree-item.js';
import './tree/util.js';
import './tree/constants.js';
import './tree/props.js';
import './tree/style';
import './tree/type.js';
import './tree-select/style';
import './tree-select/interface.js';
import './tree-select/type.js';
import './avatar/avatar.js';
import './avatar/props.js';
import './avatar/group.js';
import './avatar/avatar-group-props.js';
import './avatar/style';
import './avatar/type.js';
import './badge/badge.js';
import './badge/props.js';
import './badge/style';
import './badge/type.js';
import './calendar/calendar.js';
import './calendar/props.js';
import './calendar/utils.js';
import './calendar/const.js';
import './calendar/calendar-cell.js';
import './calendar/style';
import './calendar/type.js';
import './comment/comment.js';
import './comment/props.js';
import './comment/style';
import './comment/type.js';
import './list/list.js';
import './list/props.js';
import './list/const.js';
import './list/list-item.js';
import './list/list-item-meta.js';
import './list/list-item-meta-props.js';
import './list/style';
import './list/type.js';
import './progress/progress.js';
import './progress/constants.js';
import './progress/props.js';
import './progress/style';
import './progress/type.js';
import './skeleton/skeleton.js';
import './skeleton/props.js';
import './skeleton/style';
import './skeleton/type.js';
import './swiper/swiper.js';
import './swiper/props.js';
import './hooks/slot.js';
import './swiper/swiper-item.js';
import './swiper/style';
import './swiper/type.js';
import './table/base-table/index.js';
import 'lodash/throttle';
import './table/util/props-util.js';
import './table/base-table-props.js';
import './table/base-table/table-body.js';
import './table/base-table/table-row.js';
import './table/base-table/table-cell.js';
import './table/util/common.js';
import './table/primary-table-props.js';
import './table/base-table/table-header.js';
import './table/base-table/col-group.js';
import './table/util/interface.js';
import './hooks/virtualScroll.js';
import './table/primary-table/index.js';
import './table/primary-table/mixins/expand.js';
import './table/primary-table/expand-box.js';
import './table/primary-table/mixins/select.js';
import './table/primary-table/select-box.js';
import './table/primary-table/mixins/sort.js';
import './table/primary-table/sorter-button.js';
import './table/primary-table/mixins/row-draggable.js';
import './table/primary-table/mixins/filter.js';
import './table/primary-table/mixins/show-columns.js';
import './dialog/dialog.js';
import './dialog/actions.js';
import './dialog/props.js';
import './dialog/style';
import './dialog/type.js';
import './table/primary-table/mixins/async-loading.js';
import './table/enhanced-table/index.js';
import './table/enhanced-table-props.js';
import './table/enhanced-table/tree.js';
import './table/enhanced-table/tree-store.js';
import './table/enhanced-table/tree-select.js';
import './table/style';
import './table/type.js';
import './alert/alert.js';
import './alert/props.js';
import './hooks/icon.js';
import './alert/style';
import './alert/type.js';
import './drawer/drawer.js';
import './drawer/props.js';
import './drawer/style';
import './drawer/type.js';
import './message/style';
import './message/type.js';
import './notification/notificationList.js';
import './notification/notification.js';
import './notification/props.js';
import './notification/style';
import './notification/type.js';
import './popconfirm/popconfirm.js';
import './popconfirm/props.js';
import './popconfirm/style';
import './popconfirm/type.js';
import './upload/upload.js';
import 'lodash/findIndex';
import './upload/dragger.js';
import './upload/util.js';
import './upload/image.js';
import './upload/props.js';
import './upload/flow-list.js';
import './_common/js/upload/xhr.js';
import './upload/single-file.js';
import './upload/style';
import './upload/interface.js';
import './upload/type.js';
import './config-provider/config-provider.js';

function install(app, config) {
  Object.keys(components).forEach(function (key) {
    /plugin/i.test(key) ? app.use(components[key]) : app.use(components[key], config);
  });
}
var tdesign = {
  install: install,
  version: typeof "0.8.0" === "undefined" ? "" : "0.8.0"
};

export { tdesign as default };
