// tdesign-vue-next/lib/date-picker/basic/cell.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../../config.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var name = "".concat(config.prefix, "-date-picker-cell");
var TDatePickerCell = vue.defineComponent({
  name: name,
  props: {
    text: [String, Number],
    value: Date,
    active: Boolean,
    highlight: Boolean,
    disabled: Boolean,
    startOfRange: Boolean,
    endOfRange: Boolean,
    additional: Boolean,
    now: Boolean,
    firstDayOfMonth: Boolean,
    lastDayOfMonth: Boolean,
    onClick: Function,
    onMouseEnter: {
      type: Function
    }
  },
  render: function render() {
    var _ref;

    var text = this.text,
        value = this.value,
        active = this.active,
        highlight = this.highlight,
        disabled = this.disabled,
        startOfRange = this.startOfRange,
        endOfRange = this.endOfRange,
        additional = this.additional,
        now = this.now,
        firstDayOfMonth = this.firstDayOfMonth,
        lastDayOfMonth = this.lastDayOfMonth,
        _onClick = this.onClick,
        onMouseEnter = this.onMouseEnter;
    var cellClass = [name, (_ref = {}, _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--now"), now), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--active"), active), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--disabled"), disabled), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--highlight"), highlight), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--active-start"), startOfRange), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--active-end"), endOfRange), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--additional"), additional), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--first-day-of-month"), firstDayOfMonth), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-date-picker__cell--last-day-of-month"), lastDayOfMonth), _ref)];
    return vue.createVNode("td", {
      "class": cellClass
    }, [vue.createVNode("div", {
      "class": "".concat(config.prefix, "-date-picker__cell-wrapper"),
      "onClick": function onClick(e) {
        if (!disabled) {
          _onClick(value, {
            e: e
          });
        }
      },
      "onMouseenter": function onMouseenter() {
        return onMouseEnter && onMouseEnter(value);
      }
    }, [vue.createVNode("span", {
      "class": "".concat(config.prefix, "-date-picker__cell-text")
    }, [text])])]);
  }
});

exports["default"] = TDatePickerCell;
