// tdesign-vue-next/lib/pagination/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  current: {
    type: Number,
    "default": 1
  },
  defaultCurrent: {
    type: Number,
    "default": 1
  },
  disabled: Boolean,
  foldedMaxPageBtn: {
    type: Number,
    "default": 5
  },
  maxPageBtn: {
    type: Number,
    "default": 10
  },
  pageSize: {
    type: Number,
    "default": 10
  },
  defaultPageSize: {
    type: Number,
    "default": 10
  },
  pageSizeOptions: {
    type: Array,
    "default": function _default() {
      return [5, 10, 20, 50];
    }
  },
  showJumper: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium"].includes(val);
    }
  },
  theme: {
    type: String,
    "default": "default",
    validator: function validator(val) {
      return ["default", "simple"].includes(val);
    }
  },
  total: {
    type: Number,
    "default": 0
  },
  totalContent: {
    type: [Boolean, Function],
    "default": true
  },
  onChange: Function,
  onCurrentChange: Function,
  onPageSizeChange: Function
};

exports["default"] = props;
