// tdesign-vue-next/esm/radio/radio-button.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isVNode, defineComponent, createVNode } from 'vue';
import props from './props.js';
import _Radio from './radio.js';
import { omit } from '../utils/helper.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../config.js';
import '../utils/classnames.js';
import '../utils/event.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var _RadioButton = defineComponent({
  name: "TRadioButton",
  components: {
    Radio: _Radio
  },
  provide: function provide() {
    return {
      radioButton: this
    };
  },
  inject: {
    radioGroup: {
      "default": void 0
    }
  },
  inheritAttrs: false,
  props: _objectSpread({}, props),
  render: function render() {
    var $props = this.$props,
        $attrs = this.$attrs,
        $slots = this.$slots,
        radioGroup = this.radioGroup;
    var children = $slots["default"] && $slots["default"](null);

    var radioProps = _objectSpread(_objectSpread({}, $props), omit($attrs, Object.keys($attrs).filter(function (key) {
      return key.startsWith("on");
    })));

    if (radioGroup) {
      radioProps.checked = $props.value === radioGroup.value;
      radioProps.disabled = $props.disabled === void 0 ? radioGroup.disabled : $props.disabled;
      radioProps.name = radioGroup.name;
    }

    return createVNode(_Radio, radioProps, _isSlot(children) ? children : {
      "default": function _default() {
        return [children];
      }
    });
  }
});

export { _RadioButton as default };
