// tdesign-vue-next/lib/input-number/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  align: {
    type: String,
    validator: function validator(val) {
      return ["left", "center", "right"].includes(val);
    }
  },
  decimalPlaces: {
    type: Number,
    "default": void 0
  },
  disabled: Boolean,
  format: {
    type: Function
  },
  max: {
    type: Number,
    "default": Infinity
  },
  min: {
    type: Number,
    "default": -Infinity
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  status: {
    type: String,
    validator: function validator(val) {
      return ["success", "warning", "error"].includes(val);
    }
  },
  step: {
    type: Number,
    "default": 1
  },
  theme: {
    type: String,
    "default": "row",
    validator: function validator(val) {
      return ["column", "row", "normal"].includes(val);
    }
  },
  tips: {
    type: [String, Function]
  },
  value: {
    type: Number,
    "default": void 0
  },
  defaultValue: {
    type: Number,
    "default": void 0
  },
  onBlur: Function,
  onChange: Function,
  onEnter: Function,
  onFocus: Function,
  onKeydown: Function,
  onKeypress: Function,
  onKeyup: Function
};

exports["default"] = props;
