// tdesign-vue-next/esm/upload/dragger.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, createVNode, resolveComponent } from 'vue';
import { CheckCircleFilledIcon, ErrorCircleFilledIcon } from 'tdesign-icons-vue-next';
import { prefix } from '../config.js';
import { Loading } from '../loading/index.js';
import { Button } from '../button/index.js';
import { UPLOAD_NAME, abridgeName, returnFileSize, getCurrentDate } from './util.js';
import '../loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../utils/classnames.js';
import '../utils/dom.js';
import '../_chunks/dep-b66bfe36.js';
import 'lodash/isString';
import '../utils/easing.js';
import '../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../utils/transfer-dom.js';
import '../loading/props.js';
import '../utils/withInstall.js';
import '../loading/style';
import '../loading/type.js';
import '../loading/plugin.js';
import '../button/button.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../button/props.js';
import '../utils/ripple.js';
import '../utils/set-style.js';
import '../button/style';
import '../button/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var name = "".concat(prefix, "-upload-dragger");
var Dragger = defineComponent({
  name: name,
  components: {
    TButton: Button,
    TLoading: Loading,
    CheckCircleFilledIcon: CheckCircleFilledIcon,
    ErrorCircleFilledIcon: ErrorCircleFilledIcon
  },
  props: {
    file: {
      type: Object
    },
    loadingFile: {
      type: Object
    },
    display: {
      type: String,
      validator: function validator(val) {
        return ["file", "file-input", "image", "custom"].includes(val);
      }
    },
    cancel: Function,
    trigger: Function,
    remove: Function,
    upload: Function,
    autoUpload: Boolean
  },
  emits: ["change", "dragleave", "dragenter"],
  data: function data() {
    return {
      target: null,
      dragActive: false
    };
  },
  computed: {
    isImage: function isImage() {
      return this.display === "image";
    },
    imageUrl: function imageUrl() {
      return this.loadingFile && this.loadingFile.url || this.file && this.file.url;
    },
    percent: function percent() {
      return this.loadingFile && this.loadingFile.percent;
    },
    inputName: function inputName() {
      return this.loadingFile && this.loadingFile.name || this.file && this.file.name;
    },
    classes: function classes() {
      return ["".concat(UPLOAD_NAME, "__dragger"), _defineProperty({}, "".concat(UPLOAD_NAME, "__dragger-center"), !this.loadingFile && !this.file), _defineProperty({}, "".concat(UPLOAD_NAME, "__dragger-error"), this.loadingFile && this.loadingFile.status === "fail")];
    },
    size: function size() {
      return this.loadingFile && this.loadingFile.size || this.file && this.file.size;
    },
    showResultOperate: function showResultOperate() {
      var _this$file, _this$loadingFile;

      return Boolean(!this.loadingFile && ((_this$file = this.file) === null || _this$file === void 0 ? void 0 : _this$file.name)) || ["success", "fail"].includes((_this$loadingFile = this.loadingFile) === null || _this$loadingFile === void 0 ? void 0 : _this$loadingFile.status);
    }
  },
  methods: {
    handleDrop: function handleDrop(event) {
      event.preventDefault();
      this.$emit("change", event.dataTransfer.files);
      this.$emit("dragleave", event);
      this.dragActive = false;
    },
    handleDragenter: function handleDragenter(event) {
      this.target = event.target;
      event.preventDefault();
      this.$emit("dragenter", event);
      this.dragActive = true;
    },
    handleDragleave: function handleDragleave(event) {
      if (this.target !== event.target) return;
      event.preventDefault();
      this.$emit("dragleave", event);
      this.dragActive = false;
    },
    handleDragover: function handleDragover(event) {
      event.preventDefault();
    },
    renderDefaultDragElement: function renderDefaultDragElement() {
      var unActiveElement = createVNode("div", null, [createVNode("span", {
        "class": "".concat(prefix, "-upload--highlight")
      }, ["\u70B9\u51FB\u4E0A\u4F20"]), createVNode("span", null, ["\xA0\xA0/\xA0\xA0\u62D6\u62FD\u5230\u6B64\u533A\u57DF"])]);

      var activeElement = createVNode("div", null, ["\u91CA\u653E\u9F20\u6807"]);

      return this.dragActive ? activeElement : unActiveElement;
    },
    renderImage: function renderImage() {
      return createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__dragger-img-wrap")
      }, [this.imageUrl && createVNode("img", {
        "src": this.imageUrl
      }, null)]);
    },
    renderUploading: function renderUploading() {
      if (this.loadingFile.status === "fail") {
        return createVNode(ErrorCircleFilledIcon, null, null);
      }

      if (this.loadingFile.status === "progress") {
        return createVNode("div", {
          "class": "".concat(UPLOAD_NAME, "__single-progress")
        }, [createVNode(Loading, null, null), createVNode("span", {
          "class": "".concat(UPLOAD_NAME, "__single-percent")
        }, [Math.min(this.loadingFile.percent, 99), "%"])]);
      }
    },
    reupload: function reupload(e) {
      this.remove(e);
      this.trigger(e);
    },
    renderProgress: function renderProgress() {
      var _this$loadingFile2,
          _this$loadingFile3,
          _this = this;

      return createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__dragger-progress")
      }, [this.isImage && this.renderImage(), createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__dragger-progress-info")
      }, [createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__dragger-text")
      }, [createVNode("span", {
        "class": "".concat(UPLOAD_NAME, "__single-name")
      }, [abridgeName(this.inputName)]), this.loadingFile && this.renderUploading(), !this.loadingFile && !!this.file && createVNode(CheckCircleFilledIcon, null, null)]), createVNode("small", {
        "class": "".concat(prefix, "-size-s")
      }, ["\u6587\u4EF6\u5927\u5C0F\uFF1A", returnFileSize(this.size)]), createVNode("small", {
        "class": "".concat(prefix, "-size-s")
      }, ["\u4E0A\u4F20\u65E5\u671F\uFF1A", getCurrentDate()]), createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__dragger-btns")
      }, [["progress", "waiting"].includes((_this$loadingFile2 = this.loadingFile) === null || _this$loadingFile2 === void 0 ? void 0 : _this$loadingFile2.status) && createVNode(Button, {
        "theme": "primary",
        "variant": "text",
        "class": "".concat(UPLOAD_NAME, "__dragger-progress-cancel"),
        "onClick": this.cancel
      }, {
        "default": function _default() {
          return ["\u53D6\u6D88\u4E0A\u4F20"];
        }
      }), !this.autoUpload && ((_this$loadingFile3 = this.loadingFile) === null || _this$loadingFile3 === void 0 ? void 0 : _this$loadingFile3.status) === "waiting" && createVNode(Button, {
        "theme": "primary",
        "variant": "text",
        "onClick": function onClick(e) {
          return _this.upload(_objectSpread({}, _this.loadingFile), e);
        }
      }, {
        "default": function _default() {
          return ["\u5F00\u59CB\u4E0A\u4F20"];
        }
      })]), this.showResultOperate && createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__dragger-btns")
      }, [createVNode(resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "text",
        "class": "".concat(UPLOAD_NAME, "__dragger-progress-cancel"),
        "onClick": this.reupload
      }, {
        "default": function _default() {
          return ["\u91CD\u65B0\u4E0A\u4F20"];
        }
      }), createVNode(resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "text",
        "onClick": this.remove
      }, {
        "default": function _default() {
          return ["\u5220\u9664"];
        }
      })])])]);
    }
  },
  render: function render() {
    var content = null;

    if ((this.loadingFile || this.file) && this.display !== "custom") {
      content = this.renderProgress();
    } else {
      content = createVNode("div", {
        "class": "".concat(UPLOAD_NAME, "__trigger"),
        "onClick": this.trigger
      }, [this.$slots["default"] && this.$slots["default"](null) || this.renderDefaultDragElement()]);
    }

    return createVNode("div", {
      "class": this.classes,
      "onDrop": this.handleDrop,
      "onDragenter": this.handleDragenter,
      "onDragover": this.handleDragover,
      "onDragleave": this.handleDragleave
    }, [content]);
  }
});

export { Dragger as default };
