// tdesign-vue-next/es/anchor/anchor-target.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, createVNode, resolveComponent } from 'vue';
import { FileCopyIcon } from 'tdesign-icons-vue-next';
import { prefix } from '../config.js';
import { copyText } from '../utils/clipboard.js';
import MessagePlugin from '../message/plugin.js';
import props from './anchor-target-props.js';
import { Popup } from '../popup/index.js';
import { COMPONENT_NAME } from './constant.js';
import '../_chunks/dep-b66bfe36.js';
import '@babel/runtime/helpers/typeof';
import '../message/messageList.js';
import '../message/const.js';
import '../message/message.js';
import '../loading/index.js';
import '../loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../utils/classnames.js';
import '../utils/dom.js';
import 'lodash/isString';
import '../utils/easing.js';
import '../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../utils/transfer-dom.js';
import '../loading/props.js';
import '../utils/withInstall.js';
import './style/css.js';
import '../loading/type.js';
import '../loading/plugin.js';
import '../message/props.js';
import '../utils/event.js';
import '../utils/helper.js';
import 'lodash/camelCase';
import '../popup/popup.js';
import '@popperjs/core';
import '../popup/props.js';
import '../utils/set-style.js';
import '../utils/map-props.js';
import 'lodash/kebabCase';
import '../popup/type.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var name = "".concat(COMPONENT_NAME, "-target");
var _AnchorTarget = defineComponent({
  name: "TAnchorTarget",
  components: {
    TPopup: Popup,
    FileCopyIcon: FileCopyIcon
  },
  props: _objectSpread({}, props),
  methods: {
    copyText: function copyText$1() {
      var a = document.createElement("a");
      a.href = "#".concat(this.id);

      copyText(a.href);

      MessagePlugin.success("\u94FE\u63A5\u590D\u5236\u6210\u529F", 1e3);
    }
  },
  render: function render() {
    var _this = this;

    var TAG = this.tag,
        children = this.$slots["default"],
        id = this.id;
    var className = ["".concat(COMPONENT_NAME, "__target")];
    var iconClassName = "".concat(prefix, "-copy");
    return createVNode(TAG, {
      "id": id,
      "class": className
    }, {
      "default": function _default() {
        return [children && children(null), createVNode(resolveComponent("t-popup"), {
          "content": "\u590D\u5236\u94FE\u63A5",
          "placement": "top",
          "showArrow": true,
          "class": iconClassName
        }, {
          "default": function _default() {
            return [createVNode(FileCopyIcon, {
              "onClick": _this.copyText
            }, null)];
          }
        })];
      }
    });
  }
});

export { _AnchorTarget as default };
