// tdesign-vue-next/es/date-picker/basic/cell.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, createVNode } from 'vue';
import { prefix } from '../../config.js';

var name = "".concat(prefix, "-date-picker-cell");
var TDatePickerCell = defineComponent({
  name: name,
  props: {
    text: [String, Number],
    value: Date,
    active: Boolean,
    highlight: Boolean,
    disabled: Boolean,
    startOfRange: Boolean,
    endOfRange: Boolean,
    additional: Boolean,
    now: Boolean,
    firstDayOfMonth: Boolean,
    lastDayOfMonth: Boolean,
    onClick: Function,
    onMouseEnter: {
      type: Function
    }
  },
  render: function render() {
    var _ref;

    var text = this.text,
        value = this.value,
        active = this.active,
        highlight = this.highlight,
        disabled = this.disabled,
        startOfRange = this.startOfRange,
        endOfRange = this.endOfRange,
        additional = this.additional,
        now = this.now,
        firstDayOfMonth = this.firstDayOfMonth,
        lastDayOfMonth = this.lastDayOfMonth,
        _onClick = this.onClick,
        onMouseEnter = this.onMouseEnter;
    var cellClass = [name, (_ref = {}, _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--now"), now), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--active"), active), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--disabled"), disabled), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--highlight"), highlight), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--active-start"), startOfRange), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--active-end"), endOfRange), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--additional"), additional), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--first-day-of-month"), firstDayOfMonth), _defineProperty(_ref, "".concat(prefix, "-date-picker__cell--last-day-of-month"), lastDayOfMonth), _ref)];
    return createVNode("td", {
      "class": cellClass
    }, [createVNode("div", {
      "class": "".concat(prefix, "-date-picker__cell-wrapper"),
      "onClick": function onClick(e) {
        if (!disabled) {
          _onClick(value, {
            e: e
          });
        }
      },
      "onMouseenter": function onMouseenter() {
        return onMouseEnter && onMouseEnter(value);
      }
    }, [createVNode("span", {
      "class": "".concat(prefix, "-date-picker__cell-text")
    }, [text])])]);
  }
});

export { TDatePickerCell as default };
