// tdesign-vue-next/es/table/primary-table-props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

var primaryTableProps = {
  asyncLoading: {
    type: [String, Function]
  },
  columns: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  dragSort: Boolean,
  expandedRow: {
    type: [String, Function]
  },
  expandedRowKeys: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  defaultExpandedRowKeys: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  expandIcon: {
    type: [Boolean, Function],
    "default": true
  },
  expandOnRowClick: Boolean,
  filterIcon: {
    type: Function
  },
  filterRow: {
    type: [String, Function]
  },
  filterValue: {
    type: Object
  },
  defaultFilterValue: {
    type: Object
  },
  multipleSort: Boolean,
  selectedRowKeys: {
    type: Array
  },
  defaultSelectedRowKeys: {
    type: Array
  },
  showColumnController: Boolean,
  showDragCol: Boolean,
  sort: {
    type: [Object, Array]
  },
  defaultSort: {
    type: [Object, Array]
  },
  sortOnRowDraggable: Boolean,
  onAsyncLoadingClick: Function,
  onCellClick: Function,
  onChange: Function,
  onDataChange: Function,
  onDragSort: Function,
  onExpandChange: Function,
  onFilterChange: Function,
  onSelectChange: Function,
  onSortChange: Function
};

export { primaryTableProps as default };
