// tdesign-vue-next/es/date-picker/panel/date.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import { defineComponent, createVNode, resolveComponent } from 'vue';
import TDateHeader from '../basic/header.js';
import TDateTable from '../basic/table.js';
import { prefix } from '../../config.js';
import props from '../props.js';
import { getToday, getYears, getMonths, getWeeks, flagActive, firstUpperCase, subtractMonth, addMonth } from '../../_common/js/date-picker/utils.js';
import '@babel/runtime/helpers/defineProperty';
import 'tdesign-icons-vue-next';
import '../../button/button.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../../utils/classnames.js';
import '../../loading/index.js';
import '../../loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import '../../loading/icon/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../utils/dom.js';
import '../../_chunks/dep-b66bfe36.js';
import 'lodash/isString';
import '../../utils/easing.js';
import '../../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../../utils/transfer-dom.js';
import '../../loading/props.js';
import '../../utils/withInstall.js';
import './style/css.js';
import '../../loading/type.js';
import '../../loading/plugin.js';
import '../../button/props.js';
import '../../utils/ripple.js';
import '../../utils/set-style.js';
import '../../utils/mixins.js';
import '../../config-provider/config-receiver.js';
import '../../config-provider/zh_CN_config.js';
import '../../config-provider/type.js';
import '../basic/cell.js';
import '@babel/runtime/helpers/extends';
import 'lodash/chunk';

var name = "".concat(prefix, "-date-picker-panel");
var TDate = defineComponent({
  name: name,
  components: {
    TDateHeader: TDateHeader,
    TDateTable: TDateTable
  },
  inheritAttrs: false,
  props: {
    global: {
      type: Object,
      "default": function _default() {
        return {};
      }
    },
    value: {
      type: Date,
      "default": function _default() {
        return getToday();
      }
    },
    mode: props.mode,
    minDate: Date,
    maxDate: Date,
    firstDayOfWeek: props.firstDayOfWeek,
    disableDate: props.disableDate,
    onChange: props.onChange
  },
  emits: ["change"],
  data: function data() {
    return {
      year: this.value.getFullYear(),
      month: this.value.getMonth(),
      type: this.mode
    };
  },
  computed: {
    tableData: function tableData() {
      var year = this.year,
          month = this.month,
          type = this.type,
          value = this.value,
          mode = this.mode,
          disableDate = this.disableDate,
          minDate = this.minDate,
          maxDate = this.maxDate,
          firstDayOfWeek = this.firstDayOfWeek,
          global = this.global;
      var data;
      var options = {
        disableDate: disableDate,
        minDate: minDate,
        maxDate: maxDate,
        firstDayOfWeek: firstDayOfWeek,
        monthLocal: global.months
      };

      switch (type) {
        case "date":
          data = getWeeks({
            year: year,
            month: month
          }, options);
          break;

        case "month":
          data = getMonths(year, options);
          break;

        case "year":
          data = getYears(year, options);
          break;

        default:
          break;
      }

      var start = type === "date" || type === mode ? value : new Date(year, month);
      return flagActive(data, {
        start: start,
        type: type
      });
    }
  },
  watch: {
    mode: function mode(value) {
      this.type = value;
    },
    value: function value(_value) {
      this.year = _value.getFullYear();
      this.month = _value.getMonth();
    }
  },
  methods: {
    getClickHandler: function getClickHandler() {
      return this["click".concat(firstUpperCase(this.type))];
    },
    clickDate: function clickDate(date) {
      this.$emit("change", date);
    },
    clickMonth: function clickMonth(date) {
      if (this.mode === "month") {
        this.$emit("change", date);
      } else {
        this.type = "date";
        this.year = date.getFullYear();
        this.month = date.getMonth();
      }
    },
    clickYear: function clickYear(date) {
      if (this.mode === "year") {
        this.$emit("change", date);
      } else {
        this.type = "month";
        this.year = date.getFullYear();
      }
    },
    clickHeader: function clickHeader(flag) {
      var monthCount = 0;
      var next = null;

      switch (this.type) {
        case "date":
          monthCount = 1;
          break;

        case "month":
          monthCount = 12;
          break;

        case "year":
          monthCount = 120;
      }

      var current = new Date(this.year, this.month);

      switch (flag) {
        case 1:
          next = addMonth(current, monthCount);
          break;

        case -1:
          next = subtractMonth(current, monthCount);
          break;

        case 0:
          next = new Date();
          break;
      }

      this.year = next.getFullYear();
      this.month = next.getMonth();
    },
    onTypeChange: function onTypeChange(type) {
      this.type = type;
    }
  },
  render: function render() {
    var year = this.year,
        month = this.month,
        type = this.type,
        tableData = this.tableData,
        firstDayOfWeek = this.firstDayOfWeek;
    return createVNode("div", {
      "class": "".concat(prefix, "-date-picker__panel")
    }, [createVNode(resolveComponent("t-date-header"), {
      "year": year,
      "month": month,
      "type": type,
      "onBtnClick": this.clickHeader,
      "onTypeChange": this.onTypeChange
    }, null), createVNode(resolveComponent("t-date-table"), {
      "type": type,
      "firstDayOfWeek": firstDayOfWeek,
      "data": tableData,
      "onCellClick": this.getClickHandler()
    }, null)]);
  }
});

export { TDate as default };
