// tdesign-vue-next/lib/table/primary-table/mixins/async-loading.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var isString = require('lodash/isString');
var table_primaryTableProps = require('../../primary-table-props.js');
var loading_index = require('../../../loading/index.js');
var config = require('../../../config.js');
var utils_classnames = require('../../../utils/classnames.js');
var utils_event = require('../../../utils/event.js');
var utils_renderTnode = require('../../../utils/render-tnode.js');
require('../../../loading/loading.js');
require('@babel/runtime/helpers/slicedToArray');
require('../../../loading/icon/gradient.js');
require('../../../_common/js/loading/circle-adapter.js');
require('../../../_common/js/utils/set-style.js');
require('../../../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../../../utils/dom.js');
require('../../../_chunks/dep-eb6b0f94.js');
require('../../../utils/easing.js');
require('../../../utils/transfer-dom.js');
require('../../../loading/props.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isObject');
require('../../../utils/withInstall.js');
require('../../../loading/style');
require('../../../loading/plugin.js');
require('../../../utils/helper.js');
require('lodash/camelCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var isString__default = /*#__PURE__*/_interopDefaultLegacy(isString);

var ASYNC_LOADING_ROW = "async-loading-row";
var asyncLoadingMixin = vue.defineComponent({
  name: "".concat(config.prefix, "-primary-table-async-loading"),
  props: {
    asyncLoading: table_primaryTableProps["default"].asyncLoading
  },
  data: function data() {
    return {
      pullDownLoading: false
    };
  },
  computed: {
    classes: function classes() {
      var _ref;

      return ["".concat(config.prefix, "-table__async-loading"), (_ref = {}, _defineProperty__default["default"](_ref, utils_classnames.STATUS_CLASSNAMES.loading, this.asyncLoading === "loading"), _defineProperty__default["default"](_ref, utils_classnames.STATUS_CLASSNAMES.loadMore, this.asyncLoading === "load-more"), _ref)];
    }
  },
  methods: {
    onLoadClick: function onLoadClick() {
      if (typeof this.asyncLoading !== "string") return;
      utils_event.emitEvent(this, "async-loading-click", {
        status: this.asyncLoading
      });
    },
    renderAsyncLoadingRow: function renderAsyncLoadingRow() {
      var asyncLoadingNode = utils_renderTnode.renderTNodeJSX(this, "asyncLoading");

      if (isString__default["default"](asyncLoadingNode)) {
        var asyncLoading = this.asyncLoading;
        var loadingText = {
          "load-more": "\u70B9\u51FB\u52A0\u8F7D\u66F4\u591A",
          loading: "\u6B63\u5728\u52A0\u8F7D\u4E2D\uFF0C\u8BF7\u7A0D\u540E"
        }[String(asyncLoading)];
        return vue.createVNode("div", {
          "class": this.classes,
          "onClick": this.onLoadClick
        }, [vue.createVNode(loading_index.Loading, {
          "loading": asyncLoading === "loading",
          "text": loadingText
        }, null)]);
      }

      if (![null, false, void 0].includes(asyncLoadingNode)) {
        return vue.createVNode("div", {
          "class": this.classes,
          "onClick": this.onLoadClick
        }, [asyncLoadingNode]);
      }

      return null;
    }
  }
});

exports.ASYNC_LOADING_ROW = ASYNC_LOADING_ROW;
exports["default"] = asyncLoadingMixin;
