// tdesign-vue-next/es/date-picker/calendar-presets.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import { defineComponent, createVNode } from 'vue';
import { prefix } from '../config.js';
import { Button } from '../button/index.js';
import '../button/button.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/defineProperty';
import '../utils/classnames.js';
import '../loading/index.js';
import '../loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../utils/dom.js';
import '../_chunks/dep-b66bfe36.js';
import 'lodash/isString';
import '../utils/easing.js';
import '../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../utils/transfer-dom.js';
import '../loading/props.js';
import '../utils/withInstall.js';
import './style/css.js';
import '../loading/type.js';
import '../loading/plugin.js';
import '../button/props.js';
import '../utils/ripple.js';
import '../utils/set-style.js';
import '../button/type.js';

var TCalendarPresets = defineComponent({
  name: "TCalendarPresets",
  components: {
    TButton: Button
  },
  props: {
    global: {
      type: Object
    },
    presets: {
      type: Object
    },
    onClick: Function
  },
  methods: {
    clickPreset: function clickPreset(value) {
      this.onClick(value);
    }
  },
  render: function render() {
    var _this = this;

    var presets = this.presets;
    return createVNode("div", {
      "class": "".concat(prefix, "-date-picker__presets")
    }, [createVNode("ul", null, [presets && Object.keys(presets).map(function (key) {
      return createVNode("li", {
        "key": key
      }, [createVNode("a", {
        "onClick": function onClick() {
          return _this.clickPreset(presets[key]);
        }
      }, [key])]);
    })])]);
  }
});

export { TCalendarPresets as default };
