// tdesign-vue-next/es/select/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import mapProps from '../utils/map-props.js';
import _Select from './select.js';
import _Option from './option.js';
import _OptionGroup from './optionGroup.js';
import { withInstall } from '../utils/withInstall.js';
import './style/css.js';
import './type.js';
import '@babel/runtime/helpers/defineProperty';
import '@babel/runtime/helpers/objectWithoutProperties';
import 'vue';
import 'lodash/kebabCase';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/typeof';
import 'lodash/isFunction';
import 'lodash/debounce';
import 'lodash/get';
import 'lodash/set';
import 'tdesign-icons-vue-next';
import '../loading/index.js';
import '../loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import '../loading/icon/gradient.js';
import '../config.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../utils/classnames.js';
import '../utils/dom.js';
import '../_chunks/dep-b66bfe36.js';
import 'lodash/isString';
import '../utils/easing.js';
import '../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isObject';
import '../utils/transfer-dom.js';
import '../loading/props.js';
import '../loading/type.js';
import '../loading/plugin.js';
import '../utils/mixins.js';
import '../config-provider/config-receiver.js';
import '../config-provider/zh_CN_config.js';
import '../config-provider/type.js';
import '../input/index.js';
import '../input/addon.js';
import '../input/input.js';
import '../utils/helper.js';
import 'lodash/camelCase';
import '../input/props.js';
import '../utils/event.js';
import '../input/input-group.js';
import '../input/type.js';
import '../tag/index.js';
import '../tag/tag.js';
import '../tag/props.js';
import '../tag/check-tag.js';
import '../tag/check-tag-props.js';
import '../tag/type.js';
import '../common-components/fake-arrow.js';
import '../popup/index.js';
import '../popup/popup.js';
import '@popperjs/core';
import '../popup/props.js';
import '../utils/set-style.js';
import '../popup/type.js';
import './props.js';
import '../utils/ripple.js';
import './option-props.js';
import '../checkbox/index.js';
import '../checkbox/checkbox.js';
import '../checkbox/props.js';
import '../checkbox/group.js';
import 'lodash/intersection';
import '../checkbox/checkbox-group-props.js';
import '../checkbox/type.js';
import './option-group-props.js';

var Select = withInstall(mapProps([{
  name: "value",
  event: "change",
  alias: ["modelValue"]
}])(_Select));
var Option = withInstall(_Option);
var OptionGroup = withInstall(_OptionGroup);

export { Option, OptionGroup, Select };
