// tdesign-vue-next/lib/date-picker/date-picker.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var dayjs = require('dayjs');
var _commonjsHelpers = require('../_chunks/dep-eb6b0f94.js');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var debounce = require('lodash/debounce');
var utils_event = require('../utils/event.js');
var config = require('../config.js');
var datePicker_props = require('./props.js');
var utils_classnames = require('../utils/classnames.js');
var button_index = require('../button/index.js');
var input_index = require('../input/index.js');
var popup_index = require('../popup/index.js');
var utils_mixins = require('../utils/mixins.js');
var configProvider_configReceiver = require('../config-provider/config-receiver.js');
var datePicker_calendarPresets = require('./calendar-presets.js');
var datePicker_panel_date = require('./panel/date.js');
var datePicker_panel_dateRange = require('./panel/date-range.js');
var timePicker_panel_index = require('../time-picker/panel/index.js');
var timePicker_constant = require('../time-picker/constant.js');
var _common_js_datePicker_utils = require('../_common/js/date-picker/utils.js');
var utils_renderTnode = require('../utils/render-tnode.js');
require('../utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('lodash/camelCase');
require('../button/button.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('../utils/dom.js');
require('lodash/isString');
require('../utils/easing.js');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../utils/withInstall.js');
require('../loading/style');
require('../loading/plugin.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isObject');
require('../button/props.js');
require('../utils/ripple.js');
require('../utils/set-style.js');
require('../button/style');
require('../input/addon.js');
require('../input/input.js');
require('@babel/runtime/helpers/typeof');
require('../input/props.js');
require('../input/input-group.js');
require('../utils/map-props.js');
require('lodash/kebabCase');
require('../input/style');
require('../config-provider/zh_CN_config.js');
require('../popup/popup.js');
require('@popperjs/core');
require('../popup/props.js');
require('../popup/style');
require('./basic/header.js');
require('./basic/table.js');
require('./basic/cell.js');
require('../_chunks/dep-8baf00f6.js');
require('../time-picker/panel/props.js');
require('../time-picker/props.js');
require('@babel/runtime/helpers/extends');
require('lodash/chunk');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);
var debounce__default = /*#__PURE__*/_interopDefaultLegacy(debounce);

var isBetween$1 = {exports: {}};

(function (module, exports) {
  !function (e, i) {
    module.exports = i() ;
  }(_commonjsHelpers.commonjsGlobal, function () {

    return function (e, i, t) {
      i.prototype.isBetween = function (e, i, s, f) {
        var n = t(e),
            o = t(i),
            r = "(" === (f = f || "()")[0],
            u = ")" === f[1];
        return (r ? this.isAfter(n, s) : !this.isBefore(n, s)) && (u ? this.isBefore(o, s) : !this.isAfter(o, s)) || (r ? this.isBefore(n, s) : !this.isAfter(n, s)) && (u ? this.isAfter(o, s) : !this.isBefore(o, s));
      };
    };
  });
})(isBetween$1);

var isBetween = isBetween$1.exports;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
dayjs__default["default"].extend(isBetween);
var onOpenDebounce = debounce__default["default"](function (vm) {
  vm.createPopover();
}, 250);
var name = "".concat(config.prefix, "-date-picker");
var _DatePicker = vue.defineComponent(_objectSpread(_objectSpread({}, utils_mixins["default"](configProvider_configReceiver["default"]("datePicker"))), {}, {
  name: "TDatePicker",
  components: {
    TPopup: popup_index.Popup,
    TButton: button_index.Button,
    TInput: input_index.Input,
    TCalendarPresets: datePicker_calendarPresets["default"],
    TDate: datePicker_panel_date["default"],
    TDateRange: datePicker_panel_dateRange["default"],
    TTimePickerPanel: timePicker_panel_index["default"]
  },
  props: datePicker_props["default"],
  emits: ["input", "open", "close", "focus", "click", "change", "pick"],
  data: function data() {
    return {
      tempValue: "",
      monthDate: new Date(),
      start: new Date(),
      end: new Date(),
      selectedDates: [],
      inSelection: false,
      inline: false,
      dateFormat: "",
      multiSeparator: ",",
      inlineView: false,
      showTime: false,
      isOpen: false,
      startTimeValue: dayjs__default["default"](),
      endTimeValue: dayjs__default["default"]()
    };
  },
  computed: {
    startText: function startText() {
      return this.formatDate(this.start);
    },
    endText: function endText() {
      return this.formatDate(this.end);
    },
    formattedValue: {
      get: function get() {
        var _this = this;

        var tempValue = this.tempValue,
            range = this.range,
            mode = this.mode,
            isOpen = this.isOpen,
            startText = this.startText,
            endText = this.endText,
            global = this.global,
            outValue = this.value;
        var selectedDates = this.getDates(outValue);
        var selectedFmtDates = selectedDates.map(function (d) {
          return _this.formatDate(d);
        });

        if (tempValue) {
          return tempValue;
        }

        var strMode = range ? "range" : mode;
        var value = "";

        switch (strMode) {
          case "time":
          case "date":
          case "month":
          case "year":
            value = selectedFmtDates.join("");
            break;

          case "range":
            if (isOpen) {
              value = [startText, endText].join(global.rangeSeparator);
            } else if (selectedFmtDates.length > 1) {
              value = [selectedFmtDates[0], selectedFmtDates[1]].join(global.rangeSeparator);
            }

            break;
        }

        return value;
      },
      set: function set(value) {
        var min = this.min,
            dateFormat = this.dateFormat;

        if (value) {
          if (String(value).length >= String(this.formatDate(min || new Date())).length && dayjs__default["default"](value, dateFormat)) {
            this.tempValue = "";
            this.setDate(value, true);
          } else {
            this.tempValue = value;
          }
        }
      }
    },
    rangeText: {
      get: function get() {
        var range = this.startText;

        if (this.range) {
          range += " ".concat(this.global.rangeSeparator, " ").concat(this.endText);
        }

        return range;
      },
      set: function set(value) {
        if (this.value) {
          this.setDate(value, false);
        }
      }
    },
    min: function min() {
      var disableDate = this.disableDate || {};
      var before = disableDate.before;
      return before ? new Date(before) : null;
    },
    max: function max() {
      var disableDate = this.disableDate || {};
      var after = disableDate.after;
      return after ? new Date(after) : null;
    },
    classes: function classes() {
      var _ref;

      return [name, utils_classnames["default"].SIZE[this.size] || "", (_ref = {}, _defineProperty__default["default"](_ref, "".concat(name, "--month-picker"), this.mode === "year" || this.mode === "month"), _defineProperty__default["default"](_ref, "".concat(config.prefix, "-inline"), this.inline || this.inlineView), _ref)];
    },
    pickerStyles: function pickerStyles() {
      var _ref2;

      return _ref2 = {}, _defineProperty__default["default"](_ref2, "".concat(name, "__container"), true), _defineProperty__default["default"](_ref2, "".concat(name, "--open"), this.isOpen || this.inlineView), _defineProperty__default["default"](_ref2, "".concat(name, "--calendar-inline-view"), this.inlineView), _defineProperty__default["default"](_ref2, "".concat(name, "--range"), this.range), _ref2;
    }
  },
  mounted: function mounted() {
    this.attachDatePicker();
  },
  methods: {
    handleTimePick: function handleTimePick(col, time, index) {
      if (!this.range || index === 0) {
        var start = new Date(this.start);
        start["set".concat(_common_js_datePicker_utils.firstUpperCase(col), "s")](time);
        this.start = start;
        this.startTimeValue = dayjs__default["default"](start);
        this.dateClick(new Date(start));
      } else {
        var end = new Date(this.end);
        end["set".concat(_common_js_datePicker_utils.firstUpperCase(col), "s")](time);
        this.end = end;
        this.endTimeValue = dayjs__default["default"](end);
        this.dateClick(new Date(end));
      }
    },
    attachDatePicker: function attachDatePicker() {
      var startDate = new Date();
      var endDate = new Date();
      this.dateFormat = this.format;
      var start = new Date(startDate);
      var end = new Date(endDate);

      if (!this.range) {
        end = new Date(startDate);
      }

      this.start = start;
      this.end = end;
      var val = this.value || this.defaultValue || "";
      this.setDate(val, false);

      if (this.inlineView) {
        this.open();
      }
    },
    onNativeInput: function onNativeInput(event) {
      var val = event.target.value;
      this.formattedValue = val;
      var d1 = this.parseDate(val);

      if (d1 instanceof Date) {
        var d2 = this.formatDate(d1);
        utils_event.emitEvent(this, "input", d2);
      }
    },
    onNativeFocus: function onNativeFocus(event) {
      if (!this.isOpen) {
        this.open();
      }

      utils_event.emitEvent(this, "focus", event);
    },
    onClick: function onClick(event) {
      if (!this.isOpen) {
        this.open();
      }

      utils_event.emitEvent(this, "click", event);
    },
    normalizeDateTime: function normalizeDateTime(value, oldValue) {
      var newDate = dayjs__default["default"](value);
      var oldDate = dayjs__default["default"](oldValue);

      if (this.enableTimePicker) {
        newDate.hour(oldDate.hour());
        newDate.minute(oldDate.minute());
        newDate.second(oldDate.second());
        newDate.millisecond(oldDate.millisecond());
      }

      return newDate.toDate();
    },
    dateClick: function dateClick(value) {
      var mode = this.range ? "range" : this.mode;
      var start = this.start;
      var end = this.end;

      if (this.showTime) {
        mode = "time";
      }

      switch (mode) {
        case "time":
          this.selectedDates = this.range ? [this.start, this.end] : [value];
          this.clickedApply(false);
          break;

        case "year":
        case "month":
        case "date":
          this.start = this.normalizeDateTime(value, start);
          this.selectedDates = [this.start];
          this.clickedApply(!this.enableTimePicker);
          break;

        case "range":
          if (this.inSelection) {
            this.inSelection = false;
            this.start = this.normalizeDateTime(value[0], end);
            this.end = this.normalizeDateTime(value[1], end);

            if (this.end < this.start) {
              this.inSelection = true;
              this.start = this.normalizeDateTime(value[0], start);
            }
          } else {
            this.start = this.normalizeDateTime(value[0], start);
            this.end = this.normalizeDateTime(value[1], end);
            this.inSelection = true;
          }

          this.clickedApply(!this.enableTimePicker);
          break;
      }
    },
    toggle: function toggle() {
      if (!this.disabled) {
        if (this.isOpen) {
          this.close();
        } else {
          this.open();
        }
      }
    },
    open: function open() {
      var _this2 = this;

      if (!this.disabled) {
        var formattedValue = this.formattedValue;

        if (formattedValue) {
          this.setDate(formattedValue);
        }

        this.tempValue = "";
        this.isOpen = true;
        vue.nextTick().then(function () {
          onOpenDebounce(_this2);
          utils_event.emitEvent(_this2, "open", _this2.selectedDates);
        });
      }
    },
    close: function close() {
      if (!this.disabled) {
        this.tempValue = "";
        this.isOpen = false;
        this.showTime = false;
        utils_event.emitEvent(this, "close", this.selectedDates);
      }
    },
    clickedApply: function clickedApply() {
      var _this3 = this;

      var closePicker = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

      if (this.range) {
        this.selectedDates = [this.start, this.end];
      }

      var selectedDates = this.selectedDates.map(function (d) {
        var fd = _this3.formatDate(d);

        return fd;
      });
      this.submitInput(selectedDates, true);

      if (closePicker) {
        this.close();
      }
    },
    toggleTime: function toggleTime() {
      var _this4 = this;

      this.timeValue = dayjs__default["default"](this.start);
      this.endTimeValue = dayjs__default["default"](this.end);
      this.showTime = !this.showTime;
      this.$nextTick(function () {
        var timePickerPanel = _this4.$refs.timePickerPanel;
        timePickerPanel && timePickerPanel.panelColUpdate();
      });
    },
    clickRange: function clickRange(value) {
      if (Array.isArray(value)) {
        var _value = _slicedToArray__default["default"](value, 2),
            start = _value[0],
            end = _value[1];

        this.start = dayjs__default["default"](start).toDate();
        this.end = dayjs__default["default"](end || start).toDate();
        this.monthDate = dayjs__default["default"](start).toDate();
      } else {
        this.start = dayjs__default["default"](value).toDate();
        this.end = dayjs__default["default"](value).toDate();
        this.monthDate = dayjs__default["default"](value).toDate();
      }

      this.clickedApply();
    },
    clear: function clear() {
      var triggerChange = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
      this.close();

      if (!this.disabled) {
        var selectedDates = [];
        this.selectedDates = selectedDates;
        this.formattedValue = "";
        this.submitInput(selectedDates, triggerChange);
      }
    },
    submitInput: function submitInput(selectedDates) {
      var triggerChange = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
      var multiSeparator = this.multiSeparator;
      var mode = this.range ? "range" : this.mode;

      switch (mode) {
        case "date":
        case "month":
        case "year":
          utils_event.emitEvent(this, "input", selectedDates.join(multiSeparator));

          if (triggerChange) {
            utils_event.emitEvent(this, "change", selectedDates.join(multiSeparator));
          }

          break;

        case "range":
          utils_event.emitEvent(this, "input", selectedDates);

          if (triggerChange) {
            utils_event.emitEvent(this, "change", selectedDates);
          }

          break;
      }
    },
    parseDate: function parseDate() {
      var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var format = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "";

      if (value instanceof Date) {
        return new Date(value);
      }

      if (format) {
        var oDate = dayjs__default["default"](value, format);

        if (oDate.isValid()) {
          return new Date(oDate.toDate());
        }

        return false;
      }

      var d2 = dayjs__default["default"](value);

      if (d2.isValid()) {
        return new Date(d2.toDate());
      }

      return false;
    },
    isEnabled: function isEnabled(value) {
      var min = this.min,
          max = this.max,
          disableDate = this.disableDate,
          dateFormat = this.dateFormat;

      if (!disableDate) {
        return true;
      }

      if (typeof disableDate === "function") {
        return !disableDate(value);
      }

      if (Array.isArray(disableDate)) {
        var isIncludes = false;
        var formatedDisabledDate = disableDate.map(function (item) {
          return dayjs__default["default"](item, dateFormat);
        });
        formatedDisabledDate.forEach(function (item) {
          if (item.isSame(dayjs__default["default"](value))) {
            isIncludes = true;
          }
        });
        return !isIncludes;
      }

      if (max && min) {
        var compareMin = dayjs__default["default"](new Date(min)).startOf("day");
        var compareMax = dayjs__default["default"](new Date(max)).startOf("day");
        return dayjs__default["default"](value).isBetween(compareMin, compareMax, null, "[]");
      }

      var from = disableDate.from,
          to = disableDate.to;

      if (from && to) {
        var _compareMin = dayjs__default["default"](new Date(from)).startOf("day");

        var _compareMax = dayjs__default["default"](new Date(to)).startOf("day");

        return !dayjs__default["default"](value).isBetween(_compareMin, _compareMax, null, "[]");
      }

      return true;
    },
    setDate: function setDate() {
      var inputDate = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var triggerChange = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      if (inputDate !== 0 && !inputDate || inputDate instanceof Array && inputDate.length === 0) {
        return this.clear(triggerChange);
      }

      var selectedDates = this.getDates(inputDate);
      this.selectedDates = selectedDates;

      if (selectedDates.length > 0) {
        var _selectedDates = _slicedToArray__default["default"](selectedDates, 2),
            start = _selectedDates[0],
            end = _selectedDates[1];

        this.start = start;
        this.end = end || start;
      }
    },
    getDates: function getDates() {
      var _this5 = this;

      var inputDate = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";

      if (inputDate !== 0 && !inputDate || inputDate instanceof Array && inputDate.length === 0) {
        return [];
      }

      var format = this.dateFormat || "";
      var dates = [];

      if (inputDate instanceof Array) {
        dates = inputDate.map(function (d) {
          var d1 = _this5.parseDate(d, format);

          return d1;
        });
      } else if (inputDate instanceof Date || typeof inputDate === "number") {
        dates = [this.parseDate(inputDate, format)];
      } else if (typeof inputDate === "string") {
        var mode = this.range ? "range" : this.mode;

        switch (mode) {
          case "date":
          case "month":
          case "year":
            dates = [this.parseDate(inputDate, format)];
            break;

          case "range":
            dates = inputDate.split(this.global.rangeSeparator || "-").map(function (d) {
              var d1 = _this5.parseDate(d, format);

              return d1;
            });
            break;
        }
      }

      var selectedDates = dates.filter(function (d) {
        var isEnable = d instanceof Date && _this5.isEnabled(d);

        return isEnable;
      });
      selectedDates.sort(function (a, b) {
        return a.getTime() - b.getTime();
      });
      return selectedDates;
    },
    formatDate: function formatDate(date) {
      var format = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "";
      var dateFormat = format || this.dateFormat || this.global.format;
      var arrTime = ["H", "h", "m", "s"];
      var hasTime = arrTime.some(function (f) {
        return String(dateFormat).includes(f);
      });

      if (this.enableTimePicker && !hasTime) {
        dateFormat = [dateFormat, "HH:mm:ss"].join(" ");
      }

      var d1 = new Date(date);
      return dayjs__default["default"](d1).format(dateFormat);
    },
    createPopover: function createPopover() {
      if (this.inlineView) {
        return;
      }

      var nativeInput = this.$refs["native"];
      var tip = this.$refs.dropdownPopup;
      var refEl = nativeInput && nativeInput.$el || this.$el;

      if (!tip || !refEl) {
        return;
      }

      this.initClickAway(tip);
    },
    getPlaceholderText: function getPlaceholderText() {
      var _this$global, _this$global$placehol;

      var placeholder = this.placeholder,
          mode = this.mode;
      var placeholderStr = placeholder || ((_this$global = this.global) === null || _this$global === void 0 ? void 0 : (_this$global$placehol = _this$global.placeholder) === null || _this$global$placehol === void 0 ? void 0 : _this$global$placehol[mode]);

      if (placeholder && Array.isArray(placeholder)) {
        placeholderStr = placeholder.join(this.global.rangeSeparator);
      }

      return placeholderStr;
    }
  },
  render: function render() {
    var _this6 = this;

    var popupProps = this.popupProps,
        disabled = this.disabled,
        clearable = this.clearable,
        allowInput = this.allowInput,
        size = this.size,
        inputProps = this.inputProps,
        enableTimePicker = this.enableTimePicker,
        mode = this.mode,
        range = this.range,
        presets = this.presets,
        firstDayOfWeek = this.firstDayOfWeek;
    var start = this.start,
        end = this.end,
        showTime = this.showTime,
        startTimeValue = this.startTimeValue,
        global = this.global,
        isOpen = this.isOpen,
        endTimeValue = this.endTimeValue;
    var panelProps = {
      value: range ? [start, end] : start,
      mode: mode,
      firstDayOfWeek: firstDayOfWeek === void 0 ? 1 : firstDayOfWeek,
      disableDate: function disableDate(d) {
        return !_this6.isEnabled(d);
      },
      onChange: this.dateClick,
      global: this.global
    };

    var onPick = function onPick(date, context) {
      utils_event.emitEvent(_this6, "pick", date, context);
    };

    var panelComponent = range ? vue.createVNode(vue.resolveComponent("t-date-range"), _objectSpread(_objectSpread({}, panelProps), {}, {
      onPick: onPick
    }), null) : vue.createVNode(vue.resolveComponent("t-date"), panelProps, null);

    var popupContent = function popupContent() {
      return vue.createVNode("div", {
        "ref": "dropdownPopup",
        "class": _this6.pickerStyles
      }, [enableTimePicker && showTime && vue.createVNode("div", null, [vue.createVNode(timePicker_panel_index["default"], {
        "ref": "timePickerPanel",
        "format": _common_js_datePicker_utils.extractTimeFormat(_this6.dateFormat) || "HH:mm:ss",
        "cols": [timePicker_constant.EPickerCols.hour, timePicker_constant.EPickerCols.minute, timePicker_constant.EPickerCols.second],
        "steps": [1, 1, 1],
        "value": !range ? [startTimeValue] : [startTimeValue, endTimeValue],
        "onTimePick": _this6.handleTimePick,
        "isShowPanel": showTime,
        "isFooterDisplay": false
      }, null)]), !showTime && panelComponent, (!!presets || enableTimePicker) && vue.createVNode("div", {
        "class": "".concat(config.prefix, "-date-picker__footer")
      }, [vue.createVNode(datePicker_calendarPresets["default"], {
        "presets": presets,
        "global": global,
        "onClick": range ? _this6.clickRange : _this6.dateClick
      }, null), enableTimePicker && vue.createVNode("div", {
        "class": "".concat(name, "--apply")
      }, [enableTimePicker && vue.createVNode(vue.resolveComponent("t-button"), {
        "theme": "primary",
        "variant": "text",
        "onClick": _this6.toggleTime
      }, {
        "default": function _default() {
          return [showTime ? global.selectDate : global.selectTime];
        }
      }), vue.createVNode(vue.resolveComponent("t-button"), {
        "theme": "primary",
        "onClick": _this6.clickedApply
      }, {
        "default": function _default() {
          return [global.confirm];
        }
      })])])]);
    };

    var inputClassNames = ["".concat(config.prefix, "-form-controls"), _defineProperty__default["default"]({}, utils_classnames["default"].STATUS.active, this.isOpen)];
    var prefixIcon = utils_renderTnode.renderTNodeJSX(this, "prefixIcon");
    var suffixIconSlot = utils_renderTnode.renderTNodeJSX(this, "suffixIcon");

    var suffixIcon = function suffixIcon() {
      if (suffixIconSlot) {
        return suffixIconSlot;
      }

      if (enableTimePicker) {
        return vue.createVNode(tdesignIconsVueNext.TimeIcon, null, null);
      }

      return vue.createVNode(tdesignIconsVueNext.CalendarIcon, null, null);
    };

    return vue.createVNode("div", {
      "class": this.classes
    }, [vue.createVNode(vue.resolveComponent("t-popup"), {
      "ref": "popup",
      "class": "".concat(name, "__popup-reference"),
      "trigger": "click",
      "placement": "bottom-left",
      "disabled": disabled,
      "showArrow": false,
      "visible": isOpen,
      "popupProps": popupProps,
      "overlayClassName": name,
      "content": popupContent,
      "expandAnimation": true,
      "onVisibleChange": function onVisibleChange(visible, context) {
        if (context.trigger === "document") {
          _this6.toggle();
        }
      }
    }, {
      "default": function _default() {
        return [vue.createVNode("div", {
          "class": inputClassNames,
          "onClick": _this6.toggle
        }, [vue.createVNode(vue.resolveComponent("t-input"), {
          "ref": "native",
          "modelValue": _this6.formattedValue,
          "onUpdate:modelValue": function onUpdateModelValue($event) {
            return _this6.formattedValue = $event;
          },
          "disabled": disabled,
          "clearable": clearable,
          "placeholder": _this6.getPlaceholderText(),
          "readonly": !allowInput,
          "allowInput": allowInput ? 1 : 0,
          "size": size,
          "inputProps": inputProps,
          "onClear": function onClear(context) {
            context.e.stopPropagation();

            _this6.clear(true);
          },
          "focus": _this6.onNativeFocus,
          "input": _this6.onNativeInput,
          "click": _this6.onClick,
          "suffixIcon": suffixIcon,
          "prefixIcon": prefixIcon
        }, null)])];
      }
    })]);
  }
}));

exports["default"] = _DatePicker;
