// tdesign-vue-next/lib/date-picker/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  allowInput: Boolean,
  clearable: Boolean,
  disabled: Boolean,
  disableDate: {
    type: [Object, Array, Function]
  },
  enableTimePicker: Boolean,
  firstDayOfWeek: {
    type: Number,
    validator: function validator(val) {
      return [1, 2, 3, 4, 5, 6, 7].includes(val);
    }
  },
  format: {
    type: String,
    "default": "YYYY-MM-DD"
  },
  inputProps: {
    type: Object
  },
  mode: {
    type: String,
    "default": "month",
    validator: function validator(val) {
      return ["year", "month", "date"].includes(val);
    }
  },
  placeholder: {
    type: [String, Array]
  },
  popupProps: {
    type: Object
  },
  prefixIcon: {
    type: Function
  },
  presets: {
    type: Object
  },
  range: Boolean,
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  suffixIcon: {
    type: Function
  },
  value: {
    type: [String, Array, Date]
  },
  defaultValue: {
    type: [String, Array, Date]
  },
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  onInput: Function
};

exports["default"] = props;
