// tdesign-vue-next/lib/time-picker/constant.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../config.js');

var COMPONENT_NAME = "".concat(config.prefix, "-time-picker");
var EPickerCols = /* @__PURE__ */function (EPickerCols2) {
  EPickerCols2["hour"] = "hour";
  EPickerCols2["minute"] = "minute";
  EPickerCols2["second"] = "second";
  EPickerCols2["meridiem"] = "meridiem";
  return EPickerCols2;
}(EPickerCols || {});
var AM = "AM";
var PM = "PM";
var MERIDIEM_LIST = [AM, PM];
var amFormat = /^(a|A)\s+?[h]{1,2}(:[m]{1,2}(:[s]{1,2})?)?$/;
var pmFormat = /^[h]{1,2}(:[m]{1,2}(:[s]{1,2})?)?(\s+(a|A))?$/;
var TIME_PICKER_EMPTY = [void 0, void 0];
var KeyboardDirection = /* @__PURE__ */function (KeyboardDirection2) {
  KeyboardDirection2[KeyboardDirection2["left"] = 37] = "left";
  KeyboardDirection2[KeyboardDirection2["up"] = 38] = "up";
  KeyboardDirection2[KeyboardDirection2["right"] = 39] = "right";
  KeyboardDirection2[KeyboardDirection2["down"] = 40] = "down";
  return KeyboardDirection2;
}(KeyboardDirection || {});
var EMPTY_VALUE = -1;

exports.AM = AM;
exports.COMPONENT_NAME = COMPONENT_NAME;
exports.EMPTY_VALUE = EMPTY_VALUE;
exports.EPickerCols = EPickerCols;
exports.KeyboardDirection = KeyboardDirection;
exports.MERIDIEM_LIST = MERIDIEM_LIST;
exports.PM = PM;
exports.TIME_PICKER_EMPTY = TIME_PICKER_EMPTY;
exports.amFormat = amFormat;
exports.pmFormat = pmFormat;
