// tdesign-vue-next/lib/_chunks/dep-0e0300ae.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

var button_index = require('../button/index.js');
var divider_index = require('../divider/index.js');
var icon_index = require('../icon/index.js');
var grid_index = require('../grid/index.js');
var layout_index = require('../layout/index.js');
var affix_index = require('../affix/index.js');
var anchor_index = require('../anchor/index.js');
var breadcrumb_index = require('../breadcrumb/index.js');
var dropdown_index = require('../dropdown/index.js');
var menu_index = require('../menu/index.js');
var pagination_index = require('../pagination/index.js');
var steps_index = require('../steps/index.js');
var tabs_index = require('../tabs/index.js');
var cascader_index = require('../cascader/index.js');
var checkbox_index = require('../checkbox/index.js');
var datePicker_index = require('../date-picker/index.js');
var form_index = require('../form/index.js');
var input_index = require('../input/index.js');
var inputNumber_index = require('../input-number/index.js');
var radio_index = require('../radio/index.js');
var select_index = require('../select/index.js');
var slider_index = require('../slider/index.js');
var switch_index = require('../switch/index.js');
var tagInput_index = require('../tag-input/index.js');
var textarea_index = require('../textarea/index.js');
var transfer_index = require('../transfer/index.js');
var timePicker_index = require('../time-picker/index.js');
var treeSelect_index = require('../tree-select/index.js');
var avatar_index = require('../avatar/index.js');
var badge_index = require('../badge/index.js');
var calendar_index = require('../calendar/index.js');
var comment_index = require('../comment/index.js');
var list_index = require('../list/index.js');
var progress_index = require('../progress/index.js');
var skeleton_index = require('../skeleton/index.js');
var swiper_index = require('../swiper/index.js');
var table_index = require('../table/index.js');
var tag_index = require('../tag/index.js');
var tooltip_index = require('../tooltip/index.js');
var tree_index = require('../tree/index.js');
var alert_index = require('../alert/index.js');
var dialog_index = require('../dialog/index.js');
var drawer_index = require('../drawer/index.js');
var loading_index = require('../loading/index.js');
var message_index = require('../message/index.js');
var notification_index = require('../notification/index.js');
var popconfirm_index = require('../popconfirm/index.js');
var popup_index = require('../popup/index.js');
var upload_index = require('../upload/index.js');
var configProvider_index = require('../config-provider/index.js');
var dialog_plugin = require('../dialog/plugin.js');
var loading_plugin = require('../loading/plugin.js');
var notification_plugin = require('../notification/plugin.js');
var message_plugin = require('../message/plugin.js');
var configProvider_useReceiver = require('../config-provider/useReceiver.js');

var components = /*#__PURE__*/Object.freeze({
	__proto__: null,
	Button: button_index.Button,
	Divider: divider_index.Divider,
	Icon: icon_index.Icon,
	Row: grid_index.Row,
	Col: grid_index.Col,
	Aside: layout_index.Aside,
	Layout: layout_index.Layout,
	Header: layout_index.Header,
	Footer: layout_index.Footer,
	Content: layout_index.Content,
	Affix: affix_index.Affix,
	Anchor: anchor_index.Anchor,
	AnchorItem: anchor_index.AnchorItem,
	AnchorTarget: anchor_index.AnchorTarget,
	Breadcrumb: breadcrumb_index.Breadcrumb,
	BreadcrumbItem: breadcrumb_index.BreadcrumbItem,
	Dropdown: dropdown_index.Dropdown,
	DropdownItem: dropdown_index.DropdownItem,
	DropdownMenu: dropdown_index.DropdownMenu,
	Menu: menu_index.Menu,
	HeadMenu: menu_index.HeadMenu,
	Submenu: menu_index.Submenu,
	MenuItem: menu_index.MenuItem,
	MenuGroup: menu_index.MenuGroup,
	Pagination: pagination_index.Pagination,
	Steps: steps_index.Steps,
	StepItem: steps_index.StepItem,
	Tabs: tabs_index.Tabs,
	TabPanel: tabs_index.TabPanel,
	Cascader: cascader_index.Cascader,
	Checkbox: checkbox_index.Checkbox,
	CheckboxGroup: checkbox_index.CheckboxGroup,
	DatePicker: datePicker_index.DatePicker,
	Form: form_index.Form,
	FormItem: form_index.FormItem,
	Addon: input_index.Addon,
	Input: input_index.Input,
	InputGroup: input_index.InputGroup,
	InputNumber: inputNumber_index.InputNumber,
	Radio: radio_index.Radio,
	RadioGroup: radio_index.RadioGroup,
	RadioButton: radio_index.RadioButton,
	Select: select_index.Select,
	Option: select_index.Option,
	OptionGroup: select_index.OptionGroup,
	Slider: slider_index.Slider,
	SliderButton: slider_index.SliderButton,
	Switch: switch_index.Switch,
	TagInput: tagInput_index.TagInput,
	Textarea: textarea_index.Textarea,
	Transfer: transfer_index.Transfer,
	TimePicker: timePicker_index.TimePicker,
	TimeRangePicker: timePicker_index.TimeRangePicker,
	TreeSelect: treeSelect_index.TreeSelect,
	Avatar: avatar_index.Avatar,
	AvatarGroup: avatar_index.AvatarGroup,
	Badge: badge_index.Badge,
	Calendar: calendar_index.Calendar,
	Comment: comment_index.Comment,
	List: list_index.List,
	ListItem: list_index.ListItem,
	ListItemMeta: list_index.ListItemMeta,
	Progress: progress_index.Progress,
	Skeleton: skeleton_index.Skeleton,
	Swiper: swiper_index.Swiper,
	SwiperItem: swiper_index.SwiperItem,
	BaseTable: table_index.BaseTable,
	PrimaryTable: table_index.PrimaryTable,
	EnhancedTable: table_index.EnhancedTable,
	Tag: tag_index.Tag,
	CheckTag: tag_index.CheckTag,
	Tooltip: tooltip_index.Tooltip,
	Tree: tree_index.Tree,
	Alert: alert_index.Alert,
	DialogPlugin: dialog_plugin["default"],
	Dialog: dialog_index.Dialog,
	Drawer: drawer_index.Drawer,
	LoadingPlugin: loading_plugin["default"],
	Loading: loading_index.Loading,
	MessagePlugin: message_plugin["default"],
	Message: message_index.Message,
	NotifyPlugin: notification_plugin["default"],
	Notification: notification_index.Notification,
	Popconfirm: popconfirm_index.Popconfirm,
	Popup: popup_index.Popup,
	Upload: upload_index.Upload,
	useReceiver: configProvider_useReceiver.useReceiver,
	ConfigProvider: configProvider_index.ConfigProvider
});

exports.components = components;
