// tdesign-vue-next/es/menu/index.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _Menu from './menu.js';
import _HeadMenu from './head-menu.js';
import _Submenu from './submenu.js';
import _MenuItem from './menu-item.js';
import _MenuGroup from './menu-group.js';
import { withInstall } from '../utils/withInstall.js';
import mapProps from '../utils/map-props.js';
import './style/css.js';
import './type.js';
import '@babel/runtime/helpers/defineProperty';
import 'vue';
import '../hooks/event.js';
import '../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash/camelCase';
import '../config.js';
import './props.js';
import '../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isString';
import 'lodash/isFunction';
import 'lodash/isObject';
import './v-menu.js';
import '@babel/runtime/helpers/classCallCheck';
import '@babel/runtime/helpers/createClass';
import '@babel/runtime/helpers/toConsumableArray';
import '../_common/js/log/log.js';
import './head-menu-props.js';
import '../tabs/index.js';
import '../tabs/tabs.js';
import '../tabs/tab-panel.js';
import '../tabs/tab-panel-props.js';
import '../tabs/tab-nav.js';
import 'lodash/debounce';
import 'tdesign-icons-vue-next';
import '../tabs/tab-nav-item.js';
import '../utils/event.js';
import '../utils/ripple.js';
import '../utils/set-style.js';
import '../tabs/props.js';
import '../tabs/type.js';
import './submenu-props.js';
import '../common-components/fake-arrow.js';
import './menu-item-props.js';
import './menu-group-props.js';
import 'lodash/kebabCase';

var Menu = withInstall(mapProps([{
  name: "value",
  event: "change",
  alias: ["modelValue"]
}])(_Menu));
var HeadMenu = withInstall(mapProps([{
  name: "value",
  event: "change",
  alias: ["modelValue"]
}, {
  name: "expanded",
  event: "expand"
}])(_HeadMenu));
var Submenu = withInstall(_Submenu);
var MenuItem = withInstall(_MenuItem);
var MenuGroup = withInstall(_MenuGroup);

export { HeadMenu, Menu, MenuGroup, MenuItem, Submenu };
