// tdesign-vue-next/lib/dialog/actions.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _typeof = require('@babel/runtime/helpers/typeof');
var vue = require('vue');
var isString = require('lodash/isString');
var _isObject = require('lodash/isObject');
var utils_renderTnode = require('../utils/render-tnode.js');
var button_index = require('../button/index.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('../button/button.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../config.js');
require('../utils/classnames.js');
require('../loading/index.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/slicedToArray');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../utils/dom.js');
require('../_chunks/dep-eb6b0f94.js');
require('../utils/easing.js');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../utils/withInstall.js');
require('../loading/style');
require('../loading/plugin.js');
require('../button/props.js');
require('../utils/ripple.js');
require('../utils/set-style.js');
require('../button/style');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var isString__default = /*#__PURE__*/_interopDefaultLegacy(isString);
var _isObject__default = /*#__PURE__*/_interopDefaultLegacy(_isObject);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var ActionMixin = vue.defineComponent({
  methods: {
    getConfirmBtn: function getConfirmBtn(options) {
      var confirmBtn = options.confirmBtn,
          className = options.className;
      if (confirmBtn === null) return null;

      if (confirmBtn && this.$slots.confirmBtn) {
        console.warn("Both $props.confirmBtn and $scopedSlots.confirmBtn exist, $props.confirmBtn is preferred.");
      }

      var defaultButtonProps = this.getDefaultConfrimBtnProps(options);

      if (!confirmBtn && !this.$slots.confirmBtn) {
        return vue.createVNode(button_index.Button, vue.mergeProps({
          "class": className
        }, defaultButtonProps), null);
      }

      if (confirmBtn && ["string", "object"].includes(_typeof__default["default"](confirmBtn))) {
        return this.getButtonByProps(confirmBtn, defaultButtonProps, className);
      }

      return utils_renderTnode.renderTNodeJSX(this, "confirmBtn");
    },
    getCancelBtn: function getCancelBtn(options) {
      var cancelBtn = options.cancelBtn,
          className = options.className;
      if (cancelBtn === null) return null;

      if (cancelBtn && this.$slots.cancelBtn) {
        console.warn("Both $props.cancelBtn and $scopedSlots.cancelBtn exist, $props.cancelBtn is preferred.");
      }

      var defaultButtonProps = this.getDefaultCancelBtnProps(options);

      if (!cancelBtn && !this.$slots.cancelBtn) {
        return vue.createVNode(button_index.Button, vue.mergeProps({
          "class": className
        }, defaultButtonProps), null);
      }

      if (cancelBtn && ["string", "object"].includes(_typeof__default["default"](cancelBtn))) {
        return this.getButtonByProps(cancelBtn, defaultButtonProps);
      }

      return utils_renderTnode.renderTNodeJSX(this, "cancelBtn");
    },
    getButtonByProps: function getButtonByProps(button, defaultButton, className) {
      var newOptions = defaultButton;

      if (isString__default["default"](button)) {
        newOptions.content = button;
      } else if (_isObject__default["default"](button)) {
        newOptions = _objectSpread(_objectSpread({}, newOptions), button);
      }

      return vue.createVNode(button_index.Button, vue.mergeProps({
        "class": className
      }, newOptions), null);
    },
    confirmBtnAction: function confirmBtnAction(e) {},
    cancelBtnAction: function cancelBtnAction(e) {},
    getDefaultConfrimBtnProps: function getDefaultConfrimBtnProps(options) {
      var _this = this;

      var globalConfirm = options.globalConfirm,
          theme = options.theme,
          globalConfirmBtnTheme = options.globalConfirmBtnTheme;
      var defaultTheme = (globalConfirmBtnTheme === null || globalConfirmBtnTheme === void 0 ? void 0 : globalConfirmBtnTheme[theme]) || "primary";
      var props = {
        theme: defaultTheme,
        content: "\u786E\u5B9A",
        onClick: function onClick(e) {
          _this.confirmBtnAction(e);
        }
      };

      if (isString__default["default"](globalConfirm)) {
        props.content = globalConfirm;
      } else if (_isObject__default["default"](globalConfirm)) {
        props = _objectSpread(_objectSpread({}, props), globalConfirm);
      }

      return props;
    },
    getDefaultCancelBtnProps: function getDefaultCancelBtnProps(options) {
      var _this2 = this;

      var globalCancel = options.globalCancel;
      var props = {
        theme: "default",
        content: "\u53D6\u6D88",
        onClick: function onClick(e) {
          _this2.cancelBtnAction(e);
        }
      };

      if (isString__default["default"](globalCancel)) {
        props.content = globalCancel;
      } else if (_isObject__default["default"](globalCancel)) {
        props = _objectSpread(_objectSpread({}, props), globalCancel);
      }

      return props;
    }
  }
});

exports["default"] = ActionMixin;
