// tdesign-vue-next/lib/cascader/components/InputContent.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var isFunction = require('lodash/isFunction');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../../config.js');
var utils_classnames = require('../../utils/classnames.js');
var configProvider_configReceiver = require('../../config-provider/config-receiver.js');
var utils_mixins = require('../../utils/mixins.js');
var utils_renderTnode = require('../../utils/render-tnode.js');
var tag_index = require('../../tag/index.js');
var loading_index = require('../../loading/index.js');
var input_index = require('../../input/index.js');
var commonComponents_fakeArrow = require('../../common-components/fake-arrow.js');
var cascader_utils_inputContent = require('../utils/inputContent.js');
var cascader_utils_helper = require('../utils/helper.js');
var cascader_props = require('../props.js');
require('../../config-provider/zh_CN_config.js');
require('lodash/isEmpty');
require('lodash/isString');
require('lodash/isObject');
require('../../tag/tag.js');
require('../../tag/props.js');
require('../../utils/event.js');
require('../../utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('lodash/camelCase');
require('../../tag/check-tag.js');
require('../../tag/check-tag-props.js');
require('../../utils/map-props.js');
require('lodash/kebabCase');
require('../../utils/withInstall.js');
require('../../tag/style');
require('../../loading/loading.js');
require('../../loading/icon/gradient.js');
require('../../_common/js/loading/circle-adapter.js');
require('../../_common/js/utils/set-style.js');
require('../../_common/js/utils/helper.js');
require('../../utils/dom.js');
require('../../_chunks/dep-eb6b0f94.js');
require('../../utils/easing.js');
require('../../utils/transfer-dom.js');
require('../../loading/props.js');
require('../../loading/style');
require('../../loading/plugin.js');
require('../../input/addon.js');
require('../../input/input.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/typeof');
require('../../input/props.js');
require('../../input/input-group.js');
require('../../input/style');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var isFunction__default = /*#__PURE__*/_interopDefaultLegacy(isFunction);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var name = "".concat(config.prefix, "-cascader");
var InputContent = vue.defineComponent(_objectSpread(_objectSpread({}, utils_mixins["default"](configProvider_configReceiver["default"]("cascader"))), {}, {
  name: "".concat(name, "-input-content"),
  components: {
    Tag: tag_index.Tag,
    Input: input_index.Input,
    CloseCircleFilledIcon: tdesignIconsVueNext.CloseCircleFilledIcon
  },
  props: {
    cascaderContext: {
      type: Object
    },
    placeholder: cascader_props["default"].placeholder,
    listeners: {
      type: Object
    },
    collapsedItems: cascader_props["default"].collapsedItems
  },
  emits: ["change"],
  data: function data() {
    return {
      isHover: false
    };
  },
  computed: {
    closeIconClass: function closeIconClass() {
      return cascader_utils_inputContent.getCloseIconClass(config.prefix, utils_classnames["default"], this.cascaderContext);
    },
    fakeArrowIconClass: function fakeArrowIconClass() {
      return cascader_utils_inputContent.getFakeArrowIconClass(config.prefix, utils_classnames["default"], this.cascaderContext);
    },
    cascaderInnerClasses: function cascaderInnerClasses() {
      return cascader_utils_inputContent.getCascaderInnerClasses(config.prefix, utils_classnames["default"], this.cascaderContext);
    },
    closeShow: function closeShow() {
      return cascader_utils_inputContent.getCloseShow(this.isHover, this.cascaderContext);
    },
    singleContent: function singleContent() {
      return cascader_utils_inputContent.getSingleContent(this.cascaderContext);
    },
    multipleContent: function multipleContent() {
      return cascader_utils_inputContent.getMultipleContent(this.cascaderContext);
    },
    showPlaceholder: function showPlaceholder() {
      return cascader_utils_inputContent.getPlaceholderShow(this.cascaderContext, this.singleContent, this.multipleContent);
    }
  },
  mounted: function mounted() {
    var _this = this;

    document.addEventListener("click", function (event) {
      _this.outerClickListenerFn(event);
    });
    setTimeout(function () {
      _this.getInputWidth();
    }, 0);
  },
  unmounted: function unmounted() {
    var _this2 = this;

    document.removeEventListener("click", function (event) {
      _this2.outerClickListenerFn(event);
    });
  },
  methods: {
    getInputWidth: function getInputWidth() {
      var _this$$refs$inputCont = this.$refs.inputContent.getBoundingClientRect(),
          width = _this$$refs$inputCont.width;

      var setInputWidth = this.cascaderContext.setInputWidth;
      setInputWidth(width);
    },
    outerClickListenerFn: function outerClickListenerFn(event) {
      return cascader_utils_inputContent.outerClickListenerEffect(this.$refs.inputContent, this.cascaderContext, event);
    },
    renderContent: function renderContent() {
      var placeholder = this.placeholder,
          showPlaceholder = this.showPlaceholder;
      var content = !showPlaceholder ? this.InnerContent() : vue.createVNode("span", {
        "class": "".concat(config.prefix, "-cascader__placeholder")
      }, [placeholder || this.t(this.global.placeholder)]);
      return content;
    },
    InnerContent: function InnerContent() {
      var _this3 = this;

      var cascaderContext = this.cascaderContext,
          placeholder = this.placeholder,
          singleContent = this.singleContent,
          multipleContent = this.multipleContent,
          listeners = this.listeners,
          collapsedItems = this.collapsedItems;
      var multiple = cascaderContext.multiple,
          size = cascaderContext.size,
          disabled = cascaderContext.disabled,
          filterable = cascaderContext.filterable,
          setFilterActive = cascaderContext.setFilterActive,
          visible = cascaderContext.visible,
          inputVal = cascaderContext.inputVal,
          setInputVal = cascaderContext.setInputVal,
          minCollapsedNum = cascaderContext.minCollapsedNum,
          value = cascaderContext.value,
          showAllLevels = cascaderContext.showAllLevels;
      var _onFocus = listeners.onFocus,
          _onBlur = listeners.onBlur,
          onRemove = listeners.onRemove;

      var renderSelfTag = function renderSelfTag(node, index) {
        return vue.createVNode(tag_index.Tag, {
          "closable": !disabled,
          "key": index,
          "disabled": disabled,
          "onClose": function onClose(ctx) {
            ctx.e.stopPropagation();
            cascader_utils_inputContent.handleRemoveTagEffect(cascaderContext, node, onRemove);
          },
          "size": size
        }, {
          "default": function _default() {
            return [showAllLevels ? cascader_utils_helper.getFullPathLabel(node) : node.label];
          }
        });
      };

      var renderCollItems = function renderCollItems() {
        var tempList = [];
        multipleContent.forEach(function (node) {
          tempList.push(node.data);
        });
        return tempList;
      };

      var generalContent = !multiple ? vue.createVNode("span", {
        "class": "".concat(config.prefix, "-cascader__content")
      }, [singleContent]) : vue.createVNode("span", null, [minCollapsedNum > 0 && multipleContent.length > minCollapsedNum ? vue.createVNode("span", null, [multipleContent.slice(0, minCollapsedNum).map(function (node, index) {
        return renderSelfTag(node, index);
      }), collapsedItems || this.$slots.collapsedItems ? utils_renderTnode.renderTNodeJSX(this, "collapsedItems", {
        params: {
          value: renderCollItems(),
          collapsedSelectedItems: renderCollItems().slice(minCollapsedNum),
          count: renderCollItems().length - minCollapsedNum
        }
      }) : vue.createVNode(tag_index.Tag, {
        "size": size,
        "disabled": disabled
      }, {
        "default": function _default() {
          return ["+", multipleContent.length - minCollapsedNum];
        }
      })]) : multipleContent.map(function (node, index) {
        return renderSelfTag(node, index);
      })]);
      var inputPlaceholder = multiple ? multipleContent.map(function (node) {
        return node.label;
      }).join("\u3001") : singleContent;

      var filterContent = function filterContent() {
        return vue.createVNode(input_index.Input, {
          "size": size,
          "placeholder": inputPlaceholder || placeholder || _this3.t(_this3.locale.placeholderText),
          "value": inputVal,
          "onChange": function onChange(value2) {
            setInputVal(value2);
            setFilterActive(!!value2);
          },
          "onFocus": function onFocus(v, context) {
            return isFunction__default["default"](_onFocus) && _onFocus({
              value: value,
              e: context === null || context === void 0 ? void 0 : context.e
            });
          },
          "onBlur": function onBlur(v, context) {
            return isFunction__default["default"](_onBlur) && _onBlur({
              value: value,
              e: context === null || context === void 0 ? void 0 : context.e
            });
          },
          "autofocus": visible
        }, null);
      };

      return filterable && visible ? filterContent() : generalContent;
    },
    renderSuffixIcon: function renderSuffixIcon() {
      var _this4 = this;

      var closeShow = this.closeShow,
          closeIconClass = this.closeIconClass,
          fakeArrowIconClass = this.fakeArrowIconClass,
          _this$cascaderContext = this.cascaderContext,
          size = _this$cascaderContext.size,
          visible = _this$cascaderContext.visible,
          disabled = _this$cascaderContext.disabled,
          loading = _this$cascaderContext.loading;

      var closeIconClick = function closeIconClick(context) {
        context.e.stopPropagation();
        cascader_utils_inputContent.closeIconClickEffect(_this4.cascaderContext);
      };

      if (loading) {
        return vue.createVNode("span", {
          "class": "".concat(config.prefix, "-cascader__icon")
        }, [vue.createVNode(loading_index.Loading, {
          "size": "small"
        }, null)]);
      }

      if (closeShow) {
        return vue.createVNode(tdesignIconsVueNext.CloseCircleFilledIcon, {
          "class": closeIconClass,
          "size": size,
          "onClick": closeIconClick
        }, null);
      }

      return vue.createVNode(commonComponents_fakeArrow["default"], {
        "overlayClassName": fakeArrowIconClass,
        "isActive": visible,
        "disabled": disabled
      }, null);
    }
  },
  render: function render() {
    var _this5 = this;

    var $attrs = this.$attrs,
        cascaderContext = this.cascaderContext;
    return vue.createVNode("div", vue.mergeProps({
      "ref": "inputContent",
      "class": this.cascaderInnerClasses
    }, $attrs, {
      "onMouseenter": function onMouseenter() {
        _this5.isHover = true;
      },
      "onMouseleave": function onMouseleave() {
        _this5.isHover = false;
      },
      "onClick": function onClick(e) {
        e.preventDefault();
        cascader_utils_inputContent.innerContentClickEffect(cascaderContext);
      }
    }), [this.renderContent(), this.renderSuffixIcon()]);
  }
}));

exports["default"] = InputContent;
