// tdesign-vue-next/es/table/base-table/table-body.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isVNode, defineComponent, TransitionGroup, provide, ref, createVNode, mergeProps, resolveComponent } from 'vue';
import get from 'lodash/get';
import camelCase from 'lodash/camelCase';
import { prefix } from '../../config.js';
import TableRow from './table-row.js';
import baseTableProps from '../base-table-props.js';
import primaryTableProps from '../primary-table-props.js';
import { emitEvent } from '../../utils/event.js';
import { renderTNodeJSX } from '../../utils/render-tnode.js';
import '@babel/runtime/helpers/toConsumableArray';
import './table-cell.js';
import '../../popup/index.js';
import '../../popup/popup.js';
import '@babel/runtime/helpers/typeof';
import '@popperjs/core';
import '../../_chunks/dep-b66bfe36.js';
import '../../utils/classnames.js';
import '../../utils/dom.js';
import 'lodash/isString';
import '../../utils/easing.js';
import '../../popup/props.js';
import '../../utils/set-style.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../../utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import '../../utils/map-props.js';
import 'lodash/kebabCase';
import '../../utils/withInstall.js';
import './style/css.js';
import '../../popup/type.js';
import '../util/common.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var props = {
  data: baseTableProps.data,
  columns: baseTableProps.columns,
  rowClassName: baseTableProps.rowClassName,
  rowKey: baseTableProps.rowKey,
  rowspanAndColspan: baseTableProps.rowspanAndColspan,
  firstFullRow: baseTableProps.firstFullRow,
  lastFullRow: baseTableProps.lastFullRow,
  onCellClick: baseTableProps.onCellClick,
  rowHeight: {
    type: Number,
    "default": 0
  },
  bufferSize: {
    type: Number,
    "default": 0
  },
  scrollType: {
    type: String,
    "default": ""
  },
  handleRowMounted: {
    type: Function,
    "default": function _default() {}
  },
  trs: {
    type: Map,
    "default": function _default() {
      return /* @__PURE__ */new Map();
    }
  },
  onRowHover: baseTableProps.onRowHover,
  onRowMousedown: baseTableProps.onRowMousedown,
  onRowMouseenter: baseTableProps.onRowMouseenter,
  onRowMouseleave: baseTableProps.onRowMouseleave,
  onRowMouseup: baseTableProps.onRowMouseup,
  onRowClick: baseTableProps.onRowClick,
  onRowDbClick: baseTableProps.onRowDbClick,
  selectedRowKeys: primaryTableProps.selectedRowKeys,
  provider: {
    type: Object,
    "default": function _default() {
      return {};
    }
  },
  current: {
    type: Number,
    "default": 1
  }
};
var TableBody = defineComponent({
  name: "".concat(prefix, "-table-body"),
  components: {
    TransitionGroup: TransitionGroup
  },
  props: props,
  emits: ["row-dragstart", "row-dragover"],
  setup: function setup(props2) {
    provide("rowHeightRef", ref(props2.rowHeight));
  },
  computed: {
    selectColumn: function selectColumn() {
      return this.columns.find(function (_ref) {
        var type = _ref.type;
        return ["multiple", "single"].includes(type);
      }) || {};
    }
  },
  methods: {
    getRowspanAndColspanProps: function getRowspanAndColspanProps() {
      var props2 = [];
      var data = this.data,
          columns = this.columns,
          rowspanAndColspan = this.rowspanAndColspan;
      var cacheFirstColumnLeftedRowspan = [];
      data.forEach(function (rowData, rowIndex) {
        if (props2[rowIndex] === void 0) {
          props2[rowIndex] = {};
        }

        columns.forEach(function (col, colIndex) {
          var colKey = col.colKey;

          var _ref2 = rowspanAndColspan({
            col: col,
            colIndex: colIndex,
            row: rowData,
            rowIndex: rowIndex
          }) || {},
              rowspan = _ref2.rowspan,
              colspan = _ref2.colspan;

          rowspan = rowspan || 1;
          colspan = colspan || 1;

          if (colIndex === 0 && rowspan > 1) {
            cacheFirstColumnLeftedRowspan[rowspan + rowIndex - 1] = 1;
          }

          var leftedRowspan = 0;

          if (rowIndex === 0 || rowspan > 1) {
            leftedRowspan = rowspan - 1;
          } else {
            var _props2$preRowIndex, _props2$preRowIndex$c;

            var preRowIndex = rowIndex - 1;
            leftedRowspan = ((_props2$preRowIndex = props2[preRowIndex]) === null || _props2$preRowIndex === void 0 ? void 0 : (_props2$preRowIndex$c = _props2$preRowIndex[colKey]) === null || _props2$preRowIndex$c === void 0 ? void 0 : _props2$preRowIndex$c.leftedRowspan) || 0;

            if (leftedRowspan > 0) {
              leftedRowspan -= 1;
              rowspan = -1;
            }
          }

          var leftedColspan = 0;

          if (colIndex === 0 || colspan > 1) {
            leftedColspan = colspan - 1;

            if (colIndex === 0 && (leftedRowspan > 0 || cacheFirstColumnLeftedRowspan[rowIndex] > 0)) {
              leftedColspan = 1;
            }
          } else {
            if (rowIndex > 0) {
              var _props, _props$colKey, _props2, _props2$colKey;

              var preLeftedColspan = (_props = props2[rowIndex - 1]) === null || _props === void 0 ? void 0 : (_props$colKey = _props[colKey]) === null || _props$colKey === void 0 ? void 0 : _props$colKey.leftedColspan;
              var preLeftedColRowspan = (_props2 = props2[rowIndex - 1]) === null || _props2 === void 0 ? void 0 : (_props2$colKey = _props2[colKey]) === null || _props2$colKey === void 0 ? void 0 : _props2$colKey.leftedRowspan;

              if (preLeftedColspan > 0 && preLeftedColRowspan > 0) {
                leftedColspan = preLeftedColspan;
                colspan = -1;
              }
            }

            var preColKey = columns[colIndex - 1].colKey;

            if (leftedColspan === 0) {
              var _props2$rowIndex, _props2$rowIndex$preC;

              leftedColspan = ((_props2$rowIndex = props2[rowIndex]) === null || _props2$rowIndex === void 0 ? void 0 : (_props2$rowIndex$preC = _props2$rowIndex[preColKey]) === null || _props2$rowIndex$preC === void 0 ? void 0 : _props2$rowIndex$preC.leftedColspan) || 0;

              if (leftedColspan > 0) {
                leftedColspan -= 1;
                colspan = -1;
              }
            }
          }

          if (rowspan > 1 && colspan === -1) {
            colspan = 1;
          }

          if (colspan > 1 && rowspan === -1) {
            rowspan = 1;
          }

          props2[rowIndex][colKey] = {
            leftedColspan: leftedColspan,
            leftedRowspan: leftedRowspan,
            rowspan: rowspan,
            colspan: colspan
          };
        });
      });
      return props2;
    },
    renderFullRow: function renderFullRow(type) {
      var fullRowNode = renderTNodeJSX(this, camelCase(type));

      if (fullRowNode) {
        return createVNode("tr", null, [createVNode("td", {
          "colspan": this.columns.length,
          "class": "".concat(prefix, "-table__row--full ").concat(prefix, "-table__row-").concat(type)
        }, [fullRowNode])]);
      }

      return null;
    },
    renderBody: function renderBody() {
      var _this = this;

      var data = this.data,
          rowClassName = this.rowClassName,
          rowKey = this.rowKey,
          slots = this.$slots,
          $props = this.$props,
          rowspanAndColspan = this.rowspanAndColspan,
          selectedRowKeys = this.selectedRowKeys,
          selectColumn = this.selectColumn;
      var body = [];
      var allRowspanAndColspanProps;

      if (typeof rowspanAndColspan === "function") {
        allRowspanAndColspanProps = this.getRowspanAndColspanProps();
      }

      var columns = $props.columns,
          current = $props.current,
          scrollType = $props.scrollType,
          rowHeight = $props.rowHeight,
          bufferSize = $props.bufferSize,
          trs = $props.trs,
          provider = $props.provider,
          onRowHover = $props.onRowHover,
          onRowMouseup = $props.onRowMouseup,
          onRowMouseleave = $props.onRowMouseleave,
          onRowMouseenter = $props.onRowMouseenter,
          onRowMousedown = $props.onRowMousedown,
          onRowDbClick = $props.onRowDbClick,
          onRowClick = $props.onRowClick,
          onCellClick = $props.onCellClick;
      data.forEach(function (row, index) {
        var _row$__t_table_inner_, _this$provider$render, _this$provider;

        var defaultRowClass = typeof rowClassName === "function" ? rowClassName({
          row: row,
          rowIndex: index
        }) : rowClassName;
        var rowClass = [];

        if (defaultRowClass) {
          rowClass = rowClass.concat(defaultRowClass);
        }

        var rowspanAndColspanProps = allRowspanAndColspanProps ? allRowspanAndColspanProps[index] : void 0;
        var rowVnode;
        var key = rowKey ? get(row, rowKey) : index + _this.current;
        var disabled = typeof selectColumn.disabled === "function" ? selectColumn.disabled({
          row: row,
          rowIndex: index
        }) : selectColumn.disabled;

        if (disabled) {
          rowClass.push("".concat(prefix, "-table__row--disabled"));
        }

        if ((selectedRowKeys === null || selectedRowKeys === void 0 ? void 0 : selectedRowKeys.indexOf(key)) > -1) {
          rowClass.push("".concat(prefix, "-table__row--selected"));
        }

        if ((_row$__t_table_inner_ = row.__t_table_inner_data__) !== null && _row$__t_table_inner_ !== void 0 && _row$__t_table_inner_.level) {
          var _row$__t_table_inner_2;

          rowClass.push("".concat(prefix, "-table__row--level-").concat(((_row$__t_table_inner_2 = row.__t_table_inner_data__) === null || _row$__t_table_inner_2 === void 0 ? void 0 : _row$__t_table_inner_2.level) || 0));
        }

        var props2 = _objectSpread({
          columns: columns,
          current: current,
          scrollType: scrollType,
          rowHeight: rowHeight,
          bufferSize: bufferSize,
          trs: trs,
          provider: provider,
          onRowHover: onRowHover,
          onRowMouseup: onRowMouseup,
          onRowMouseleave: onRowMouseleave,
          onRowMouseenter: onRowMouseenter,
          onRowMousedown: onRowMousedown,
          onRowDbClick: onRowDbClick,
          onRowClick: onRowClick,
          onCellClick: onCellClick,
          key: key,
          rowClass: rowClass.join(" "),
          rowData: row,
          index: index,
          rowspanAndColspanProps: rowspanAndColspanProps
        }, {
          onRowDragstart: function onRowDragstart() {
            emitEvent(_this, "row-dragstart", {
              index: index,
              data: row
            });
          },
          onRowDragover: function onRowDragover(_ref3) {
            var e = _ref3.e;
            e.preventDefault();
            emitEvent(_this, "row-dragover", {
              index: index,
              data: row,
              targetElm: rowVnode.el
            });
          }
        });

        rowVnode = createVNode(TableRow, mergeProps({
          "rowKey": _this.rowKey
        }, props2, {
          "onRowMounted": _this.handleRowMounted,
          "row": row
        }), _isSlot(slots) ? slots : {
          "default": function _default() {
            return [slots];
          }
        });
        body.push(rowVnode);
        var expandedRow = (_this$provider$render = (_this$provider = _this.provider).renderExpandedRow) === null || _this$provider$render === void 0 ? void 0 : _this$provider$render.call(_this$provider, {
          row: row,
          index: index
        });
        expandedRow && (body = body.concat(expandedRow));
      });
      var firstRow = this.renderFullRow("first-full-row");

      if (firstRow) {
        body = [firstRow].concat(body);
      }

      var lastRow = this.renderFullRow("last-full-row");

      if (lastRow) {
        body = body.concat(lastRow);
      }

      return body;
    }
  },
  render: function render() {
    if (this.provider.sortOnRowDraggable) {
      var _slot;

      var className = "".concat(prefix, "-table__body ").concat(this.provider.dragging ? "".concat(prefix, "-table__body--dragging") : "");
      return createVNode(resolveComponent("transition-group"), {
        "class": className,
        "tag": "tbody"
      }, _isSlot(_slot = this.renderBody()) ? _slot : {
        "default": function _default() {
          return [_slot];
        }
      });
    }

    return createVNode("tbody", {
      "class": "".concat(prefix, "-table__body")
    }, [this.renderBody()]);
  }
});

export { TableBody as default };
