// tdesign-vue-next/lib/table/base-table-props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var baseTableProps = {
  bordered: Boolean,
  columns: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  data: {
    type: Array,
    "default": function _default() {
      return [];
    }
  },
  disableDataSort: Boolean,
  empty: {
    type: [String, Function],
    "default": ""
  },
  firstFullRow: {
    type: [String, Function]
  },
  height: {
    type: [String, Number],
    "default": "auto"
  },
  hover: Boolean,
  lastFullRow: {
    type: [String, Function]
  },
  loading: {
    type: [Boolean, Function],
    "default": false
  },
  maxHeight: {
    type: [String, Number]
  },
  pagination: {
    type: Object
  },
  rowClassName: {
    type: [String, Object, Array, Function]
  },
  rowKey: {
    type: String,
    "default": "",
    required: true
  },
  rowspanAndColspan: {
    type: Function
  },
  scroll: {
    type: Object
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  stripe: Boolean,
  tableLayout: {
    type: String,
    "default": "fixed",
    validator: function validator(val) {
      return ["auto", "fixed"].includes(val);
    }
  },
  topContent: {
    type: [String, Function]
  },
  verticalAlign: {
    type: String,
    "default": "middle",
    validator: function validator(val) {
      return ["top", "middle", "bottom"].includes(val);
    }
  },
  onCellClick: Function,
  onPageChange: Function,
  onRowClick: Function,
  onRowDbClick: Function,
  onRowHover: Function,
  onRowMousedown: Function,
  onRowMouseenter: Function,
  onRowMouseleave: Function,
  onRowMouseup: Function,
  onScrollX: Function,
  onScrollY: Function
};

exports["default"] = baseTableProps;
