// tdesign-vue-next/lib/comment/comment.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var config = require('../config.js');
var comment_props = require('./props.js');
var utils_renderTnode = require('../utils/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isString');
require('lodash/isFunction');
require('lodash/isObject');

var preName = "".concat(config.prefix, "-comment");
var _Comment = vue.defineComponent({
  name: "TComment",
  props: comment_props["default"],
  methods: {
    renderReply: function renderReply() {
      var reply = utils_renderTnode.renderTNodeJSX(this, "reply");
      return reply ? vue.createVNode("div", {
        "class": "".concat(preName, "__reply")
      }, [reply]) : null;
    },
    renderActions: function renderActions() {
      var actions = utils_renderTnode.renderTNodeJSX(this, "actions");
      return actions && actions.length ? vue.createVNode("ul", {
        "class": "".concat(preName, "__actions")
      }, [actions.map(function (action, index) {
        return vue.createVNode("li", {
          "key": "action-".concat(index)
        }, [action]);
      })]) : null;
    },
    renderQuote: function renderQuote() {
      var quote = utils_renderTnode.renderTNodeJSX(this, "quote");
      return quote ? vue.createVNode("div", {
        "class": "".concat(preName, "__quote")
      }, [quote]) : null;
    },
    renderAuthorDatetime: function renderAuthorDatetime() {
      var author = utils_renderTnode.renderTNodeJSX(this, "author");
      var datetime = utils_renderTnode.renderTNodeJSX(this, "datetime");
      return (author || datetime) && vue.createVNode("div", {
        "class": "".concat(preName, "__author")
      }, [author && vue.createVNode("span", {
        "class": "".concat(preName, "__name")
      }, [author]), datetime && vue.createVNode("span", {
        "class": "".concat(preName, "__time")
      }, [datetime])]);
    },
    renderContent: function renderContent() {
      return vue.createVNode("div", {
        "class": "".concat(preName, "__content")
      }, [this.renderAuthorDatetime(), vue.createVNode("div", {
        "class": "".concat(preName, "__detail")
      }, [utils_renderTnode.renderTNodeJSX(this, "content")]), this.renderQuote(), this.renderActions()]);
    },
    renderAvatar: function renderAvatar() {
      var avatar = utils_renderTnode.renderTNodeJSX(this, "avatar");
      return avatar ? vue.createVNode("div", {
        "class": "".concat(preName, "__avatar")
      }, [typeof avatar === "string" ? vue.createVNode("img", {
        "src": avatar,
        "alt": "",
        "class": "".concat(preName, "__avatar-image")
      }, null) : avatar]) : null;
    }
  },
  render: function render() {
    return vue.createVNode("div", {
      "class": preName
    }, [vue.createVNode("div", {
      "class": "".concat(preName, "__inner")
    }, [this.renderAvatar(), this.renderContent()]), this.renderReply()]);
  }
});

exports["default"] = _Comment;
