// tdesign-vue-next/esm/table/primary-table/mixins/show-columns.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import { isVNode, defineComponent, createVNode } from 'vue';
import { BulletpointIcon } from 'tdesign-icons-vue-next';
import { Button } from '../../../button/index.js';
import { Dialog } from '../../../dialog/index.js';
import { Checkbox } from '../../../checkbox/index.js';
import CheckboxGroup from '../../../checkbox/group.js';
import primaryTableProps from '../../primary-table-props.js';
import { prefix } from '../../../config.js';
import '../../../button/button.js';
import '@babel/runtime/helpers/defineProperty';
import '../../../utils/classnames.js';
import '../../../loading/index.js';
import '../../../loading/loading.js';
import '@babel/runtime/helpers/slicedToArray';
import '../../../loading/icon/gradient.js';
import '../../../_common/js/loading/circle-adapter.js';
import '../../../_common/js/utils/set-style.js';
import '../../../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../../utils/dom.js';
import '../../../_chunks/dep-b66bfe36.js';
import 'lodash/isString';
import '../../../utils/easing.js';
import '../../../utils/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isFunction';
import 'lodash/isObject';
import '../../../utils/transfer-dom.js';
import '../../../loading/props.js';
import '../../../utils/withInstall.js';
import '../../../loading/style';
import '../../../loading/type.js';
import '../../../loading/plugin.js';
import '../../../button/props.js';
import '../../../utils/ripple.js';
import '../../../utils/set-style.js';
import '../../../button/style';
import '../../../button/type.js';
import '../../../dialog/dialog.js';
import '../../../dialog/actions.js';
import '@babel/runtime/helpers/typeof';
import '../../../dialog/props.js';
import '../../../utils/event.js';
import '../../../utils/helper.js';
import 'lodash/camelCase';
import '../../../utils/mixins.js';
import '../../../config-provider/config-receiver.js';
import '../../../config-provider/zh_CN_config.js';
import '../../../config-provider/type.js';
import '../../../dialog/style';
import '../../../dialog/type.js';
import '../../../dialog/plugin.js';
import '../../../checkbox/checkbox.js';
import '../../../checkbox/props.js';
import '../../../utils/map-props.js';
import 'lodash/kebabCase';
import '../../../checkbox/style';
import '../../../checkbox/type.js';
import 'lodash/intersection';
import '../../../checkbox/checkbox-group-props.js';

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var content = "\u81EA\u5B9A\u4E49\u5217";
var showColumns = defineComponent({
  name: "".concat(prefix, "-primary-show-columns"),
  components: {
    Button: Button,
    BulletpointIcon: BulletpointIcon,
    Dialog: Dialog,
    Checkbox: Checkbox,
    CheckboxGroup: CheckboxGroup
  },
  props: {
    columns: primaryTableProps.columns,
    showColumns: {
      type: Boolean,
      "default": false
    }
  },
  data: function data() {
    return {
      showColumnKeys: this.columns.map(function (_ref) {
        var colKey = _ref.colKey;
        return colKey;
      }),
      showColumnCheckboxKeys: this.columns.map(function (_ref2) {
        var colKey = _ref2.colKey;
        return colKey;
      }),
      isShowColumnsDlg: false
    };
  },
  computed: {
    showColumnCheckboxOpts: function showColumnCheckboxOpts() {
      return this.columns.filter(function (_ref3) {
        var colKey = _ref3.colKey,
            title = _ref3.title;
        return colKey && title;
      }).map(function (_ref4) {
        var colKey = _ref4.colKey,
            title = _ref4.title;
        return {
          label: title,
          value: colKey
        };
      });
    },
    isAllShowColumns: function isAllShowColumns() {
      var _this = this;

      return this.showColumnCheckboxOpts.every(function (_ref5) {
        var value = _ref5.value;
        return _this.showColumnCheckboxKeys.includes(value);
      });
    },
    isSomeShowColumns: function isSomeShowColumns() {
      var _this2 = this;

      return !this.isAllShowColumns && this.showColumnCheckboxOpts.some(function (_ref6) {
        var value = _ref6.value;
        return _this2.showColumnCheckboxKeys.includes(value);
      });
    }
  },
  watch: {
    columns: {
      deep: true,
      handler: function handler() {
        this.updateColumns();
      }
    }
  },
  methods: {
    getShowColumns: function getShowColumns(columns) {
      var _this3 = this;

      return columns.filter(function (_ref7) {
        var colKey = _ref7.colKey;
        return _this3.showColumnKeys.includes(colKey) || !_this3.showColumnCheckboxOpts.map(function (_ref8) {
          var value = _ref8.value;
          return value;
        }).includes(colKey);
      });
    },
    renderShowColumnsDlgFooter: function renderShowColumnsDlgFooter() {
      return createVNode("div", null, [createVNode(Button, {
        "variant": "base",
        "theme": "primary",
        "onClick": this.handleConfirmShowColumnsDlg
      }, {
        "default": function _default() {
          return ["\u786E\u5B9A"];
        }
      }), createVNode(Button, {
        "onClick": this.handleConfirmShowColumnsDlg
      }, {
        "default": function _default() {
          return ["\u53D6\u6D88"];
        }
      })]);
    },
    renderShowColumnsDlg: function renderShowColumnsDlg() {
      var _this4 = this;

      return createVNode(Dialog, {
        "header": content,
        "footer": this.renderShowColumnsDlgFooter,
        "visible": this.isShowColumnsDlg
      }, {
        "default": function _default() {
          return [createVNode("div", {
            "slot": "body"
          }, [createVNode("div", null, [createVNode(Checkbox, {
            "indeterminate": _this4.isSomeShowColumns,
            "checked": _this4.isAllShowColumns,
            "onChange": _this4.handleClickAllShowColumns
          }, {
            "default": function _default() {
              return ["\u5168\u9009"];
            }
          })]), createVNode(CheckboxGroup, {
            "options": _this4.showColumnCheckboxOpts,
            "modelValue": _this4.showColumnCheckboxKeys,
            "onUpdate:modelValue": function onUpdateModelValue($event) {
              return _this4.showColumnCheckboxKeys = $event;
            }
          }, null)])];
        }
      });
    },
    renderShowColumns: function renderShowColumns() {
      var _this5 = this;

      return createVNode("div", null, [createVNode("div", null, [createVNode(Button, {
        "icon": function icon() {
          return createVNode(BulletpointIcon, null, null);
        },
        "onClick": function onClick() {
          return _this5.isShowColumnsDlg = true;
        }
      }, _isSlot(content) ? content : {
        "default": function _default() {
          return [content];
        }
      })]), this.renderShowColumnsDlg()]);
    },
    handleClickAllShowColumns: function handleClickAllShowColumns() {
      this.showColumnCheckboxKeys = this.isAllShowColumns ? [] : this.showColumnCheckboxOpts.map(function (_ref9) {
        var value = _ref9.value;
        return value;
      });
    },
    handleConfirmShowColumnsDlg: function handleConfirmShowColumnsDlg() {
      this.showColumnKeys = _toConsumableArray(this.showColumnCheckboxKeys);
      this.handleCancelShowColumnsDlg();
    },
    handleCancelShowColumnsDlg: function handleCancelShowColumnsDlg() {
      this.isShowColumnsDlg = false;
    },
    updateColumns: function updateColumns() {
      this.showColumnKeys = this.columns.map(function (_ref10) {
        var colKey = _ref10.colKey;
        return colKey;
      });
      this.showColumnCheckboxKeys = this.columns.map(function (_ref11) {
        var colKey = _ref11.colKey;
        return colKey;
      });
    }
  }
});

export { showColumns as default };
