// tdesign-vue-next/lib/grid/col.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../config.js');
var grid_colProps = require('./col-props.js');
var utils_renderTnode = require('../utils/render-tnode.js');
var grid_common = require('./common.js');
require('lodash/isEmpty');
require('lodash/isString');
require('lodash/isFunction');
require('lodash/isObject');
require('../utils/responsive.js');
require('../hooks/event.js');
require('../utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('lodash/camelCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !vue.isVNode(s);
}

var name = "".concat(config.prefix, "-col");
var _Col = vue.defineComponent({
  name: "TCol",
  inject: ["rowContext"],
  props: _objectSpread({}, grid_colProps["default"]),
  setup: function setup(props2) {
    var rowContext = vue.inject("rowContext", /* @__PURE__ */Object.create(null));
    var size = grid_common.useRowSize();
    var colClasses = vue.computed(function () {
      return grid_common.getColClasses(name, props2);
    });
    var colStyle = vue.computed(function () {
      var colStyle2 = {};
      var flex = props2.flex;
      flex && (colStyle2.flex = grid_common.parseFlex(flex));

      if (rowContext) {
        var rowGutter = rowContext.gutter;
        Object.assign(colStyle2, grid_common.calcColPadding(rowGutter, size.value));
      }

      return colStyle2;
    });
    return {
      size: size,
      colClasses: colClasses,
      colStyle: colStyle
    };
  },
  render: function render() {
    var _slot;

    var colStyle = this.colStyle,
        TAG = this.tag,
        colClasses = this.colClasses;
    return vue.createVNode(TAG, {
      "class": colClasses,
      "style": colStyle
    }, _isSlot(_slot = utils_renderTnode.renderTNodeJSX(this, "default")) ? _slot : {
      "default": function _default() {
        return [_slot];
      }
    });
  }
});

exports["default"] = _Col;
