// tdesign-vue-next/lib/popup/props.js
/**
 * tdesign v0.8.0
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var popupProps = {
  attach: {
    type: [String, Function],
    "default": "body"
  },
  content: {
    type: [String, Function]
  },
  "default": {
    type: [String, Function]
  },
  destroyOnClose: Boolean,
  disabled: Boolean,
  overlayClassName: {
    type: [String, Object, Array]
  },
  overlayStyle: {
    type: [Boolean, Object, Function]
  },
  placement: {
    type: String,
    "default": "top",
    validator: function validator(val) {
      return ["top", "left", "right", "bottom", "top-left", "top-right", "bottom-left", "bottom-right", "left-top", "left-bottom", "right-top", "right-bottom"].includes(val);
    }
  },
  showArrow: Boolean,
  trigger: {
    type: String,
    "default": "hover",
    validator: function validator(val) {
      return ["hover", "click", "focus", "context-menu"].includes(val);
    }
  },
  triggerElement: {
    type: [String, Function]
  },
  visible: Boolean,
  defaultVisible: Boolean,
  zIndex: {
    type: Number
  },
  onVisibleChange: Function
};

exports["default"] = popupProps;
